% EN: ﻿\chapter{ScopeGuard}
\Trans{﻿\chapter{ScopeGuard}}{}

% EN: This chapter covers a pattern that can be seen as a generalization of the RAII idiom we studied earlier. In its earliest form, it is an old and established C++ pattern, however, it is also one that has particularly benefited from the language additions in C++11, C++14, and C++17. We will witness the evolution of this pattern as the language becomes more powerful. The ScopeGuard pattern exists at the intersection of declarative programming (say what you want to happen, not how you want it done) and error-safe programs (especially exception safety). We will have to learn a bit about both before we fully understand ScopeGuard.
\Trans{This chapter covers a pattern that can be seen as a generalization of the RAII idiom we studied earlier. In its earliest form, it is an old and established C++ pattern, however, it is also one that has particularly benefited from the language additions in C++11, C++14, and C++17. We will witness the evolution of this pattern as the language becomes more powerful. The ScopeGuard pattern exists at the intersection of declarative programming (say what you want to happen, not how you want it done) and error-safe programs (especially exception safety). We will have to learn a bit about both before we fully understand ScopeGuard.}{}

% EN: The following topics will be covered in this chapter:
\Trans{The following topics will be covered in this chapter:}{}

\begin{itemize}
\item
% EN:   How can we write error-safe and exception-safe code? How does RAII make error handling easier?
\Trans{How can we write error-safe and exception-safe code? How does RAII make error handling easier?}{}
\item
% EN:   What is composability as applied to error handling?
\Trans{What is composability as applied to error handling?}{}
\item
% EN:   Why is RAII not powerful enough for error handling, and how is it generalized? How can we implement declarative error handling in C++?
\Trans{Why is RAII not powerful enough for error handling, and how is it generalized? How can we implement declarative error handling in C++?}{}
\end{itemize}

\section{Technical requirements}

% EN: Here is the example code: https://github.com/PacktPublishing/Hands-On-Design-Patterns-with-CPP-Second-Edition/tree/master/Chapter11.
\Trans{Here is the example code: https://github.com/PacktPublishing/Hands-On-Design-Patterns-with-CPP-Second-Edition/tree/master/Chapter11.}{}

% EN: You will also need the Google Benchmark library installed and configured: https://github.com/google/benchmark (see \emph{Chapter 4, Swap -- From Simple to Subtle}, for installation instructions).
\Trans{You will also need the Google Benchmark library installed and configured: https://github.com/google/benchmark (see \emph{Chapter 4, Swap -- From Simple to Subtle}, for installation instructions).}{}

% EN: This chapter is rather heavy on advanced C++ features, so keep a C++ reference nearby (https://en.cppreference.com unless you want to dig through the standard itself).
\Trans{This chapter is rather heavy on advanced C++ features, so keep a C++ reference nearby (https://en.cppreference.com unless you want to dig through the standard itself).}{}

% EN: Finally, a very thorough and complete implementation of ScopeGuard can be found in the Folly library: https://github.com/facebook/folly/blob/master/folly/ScopeGuard.h{}; it includes C++ library programming details beyond those covered in this book.
\Trans{Finally, a very thorough and complete implementation of ScopeGuard can be found in the Folly library: https://github.com/facebook/folly/blob/master/folly/ScopeGuard.h{}; it includes C++ library programming details beyond those covered in this book.}{}

\section{Error handling and resource acquisition is initialization}

% EN: We begin by reviewing the concepts of error handling, and, in particular, writing exception-safe code in C++. The \textbf{Resource Acquisition Is Initialization} (\textbf{RAII}) idiom is one of the primary methods of error handling in C++. We have already dedicated an entire chapter to it, and you will need it here to make sense of what we are about to do. Let's first recognize the problem we are facing.
\Trans{We begin by reviewing the concepts of error handling, and, in particular, writing exception-safe code in C++. The \textbf{Resource Acquisition Is Initialization} (\textbf{RAII}) idiom is one of the primary methods of error handling in C++. We have already dedicated an entire chapter to it, and you will need it here to make sense of what we are about to do. Let's first recognize the problem we are facing.}{}

\subsection{Error safety and exception safety}

% EN: For the rest of this chapter, we will consider the following problem---suppose we are implementing a database of records. The records are stored on disk, but there is also an in-memory index for fast access to the records. The database API offers a method to insert records into the database:
\Trans{For the rest of this chapter, we will consider the following problem---suppose we are implementing a database of records. The records are stored on disk, but there is also an in-memory index for fast access to the records. The database API offers a method to insert records into the database:}{}

\begin{code}
% EN: class Record { ...  };
class Database {
  public:
  void insert(const Record& r);
  ...
 };
\Trans{class Record { ... }; class Database { public: void insert(const Record& r); ... };}{}

\end{code}

% EN: If the insertion succeeds, both the index and the disk storage are updated and consistent with each other. If something goes wrong, an exception is thrown.
\Trans{If the insertion succeeds, both the index and the disk storage are updated and consistent with each other. If something goes wrong, an exception is thrown.}{}

% EN: While it appears to the clients of the database that the insertion is a single transaction, the implementation has to deal with the fact that it is done in multiple steps---we need to insert the record into the index and write it to disk. To facilitate this, the database contains two classes, each responsible for its type of storage:
\Trans{While it appears to the clients of the database that the insertion is a single transaction, the implementation has to deal with the fact that it is done in multiple steps---we need to insert the record into the index and write it to disk. To facilitate this, the database contains two classes, each responsible for its type of storage:}{}

\begin{code}
% EN: class Database {
  class Storage { ...  };    // Disk storage Storage S;
  class Index { ...  };    // Memory index Index I;
  public:
  void insert(const Record& r);
  ...
 };
\Trans{class Database { class Storage { ... }; // Disk storage Storage S; class Index { ... }; // Memory index Index I; public: void insert(const Record& r); ... };}{}

\end{code}

% EN: The implementation of the \texttt{insert()} function must insert the record into both the storage and the index, there is no way around it:
\Trans{The implementation of the \texttt{insert()} function must insert the record into both the storage and the index, there is no way around it:}{}

\begin{code}
% EN: //Example 01
void Database::insert(const Record& r) {
  S.insert(r);
  I.insert(r);
\Trans{//Example 01 void Database::insert(const Record& r) { S.insert(r); I.insert(r);}{}

}
\end{code}

% EN: Unfortunately, either of these operations can fail. Let's see first what happens if the storage insertion fails. Let's assume that all failures in the program are signaled by throwing exceptions. If the storage insertion fails, the storage remains unchanged, the index insertion is not attempted at all, and the exception is propagated out of the \texttt{Database::insert()} function. That is exactly what we want---the insertion failed, the database is unchanged, and an exception is thrown.
\Trans{Unfortunately, either of these operations can fail. Let's see first what happens if the storage insertion fails. Let's assume that all failures in the program are signaled by throwing exceptions. If the storage insertion fails, the storage remains unchanged, the index insertion is not attempted at all, and the exception is propagated out of the \texttt{Database::insert()} function. That is exactly what we want---the insertion failed, the database is unchanged, and an exception is thrown.}{}

% EN: So, what happens if the storage insertion succeeds but the index fails? Things are not looking so good this time---the disk is altered successfully, then the index insertion fails, and the exception propagates to the caller of \texttt{Database::insert()} to signal the insertion failure, but the truth is the insertion did not completely fail. It did not completely succeed either.
\Trans{So, what happens if the storage insertion succeeds but the index fails? Things are not looking so good this time---the disk is altered successfully, then the index insertion fails, and the exception propagates to the caller of \texttt{Database::insert()} to signal the insertion failure, but the truth is the insertion did not completely fail. It did not completely succeed either.}{}

% EN: The database is left in an inconsistent state; there is a record on disk that is not accessible from the index. This is a failure to handle an error condition, exception-unsafe code, and it just won't do.
\Trans{The database is left in an inconsistent state; there is a record on disk that is not accessible from the index. This is a failure to handle an error condition, exception-unsafe code, and it just won't do.}{}

% EN: The impulsive attempt to change the order of the sub-operations does not help:
\Trans{The impulsive attempt to change the order of the sub-operations does not help:}{}

\begin{code}
% EN: void Database::insert(const Record& r) {
  I.insert(r);
  S.insert(r);
\Trans{void Database::insert(const Record& r) { I.insert(r); S.insert(r);}{}

}
\end{code}

% EN: Sure, everything is now fine if the index fails. But we have the same problem if the storage insertion throws an exception---we now have an entry in the index that is pointing to nowhere since the record was never written to disk.
\Trans{Sure, everything is now fine if the index fails. But we have the same problem if the storage insertion throws an exception---we now have an entry in the index that is pointing to nowhere since the record was never written to disk.}{}

% EN: Obviously, we cannot just ignore the exceptions thrown by \texttt{Index} or \texttt{Storage}; we have to somehow deal with them to maintain the consistency of the database. We know how to handle exceptions; that's what the \texttt{try-catch} block is for:
\Trans{Obviously, we cannot just ignore the exceptions thrown by \texttt{Index} or \texttt{Storage}; we have to somehow deal with them to maintain the consistency of the database. We know how to handle exceptions; that's what the \texttt{try-catch} block is for:}{}

\begin{code}
% EN: // Example 02
void Database::insert(const Record& r) {
  S.insert(r);
  try {
    I.insert(r);
   } catch (...) {
    S.undo();
    throw;    // Rethrow
   }
\Trans{// Example 02 void Database::insert(const Record& r) { S.insert(r); try { I.insert(r); } catch (...) { S.undo(); throw; // Rethrow }}{}

}
\end{code}

% EN: Again, if the storage fails, we don't need to do anything special. If the index fails, we have to undo the last operation on the storage (let's assume that it has the API to do that). Now the database is again consistent as if the insertion never happened. Even though we caught the exception thrown by the index, we still need to signal to the caller that the insertion fails, so we re-throw the exception. So far, so good.
\Trans{Again, if the storage fails, we don't need to do anything special. If the index fails, we have to undo the last operation on the storage (let's assume that it has the API to do that). Now the database is again consistent as if the insertion never happened. Even though we caught the exception thrown by the index, we still need to signal to the caller that the insertion fails, so we re-throw the exception. So far, so good.}{}

% EN: The situation is not much different if we choose to use error codes instead of exceptions; let's consider the variant where all \texttt{insert()} functions return \texttt{true} if they succeed and \texttt{false} if they fail:
\Trans{The situation is not much different if we choose to use error codes instead of exceptions; let's consider the variant where all \texttt{insert()} functions return \texttt{true} if they succeed and \texttt{false} if they fail:}{}

\begin{code}
% EN: bool Database::insert(const Record& r) {
  if (!S.insert(r)) return false;
  if (!I.insert(r)) {
    S.undo();
    return false;
   }
  return true;
\Trans{bool Database::insert(const Record& r) { if (!S.insert(r)) return false; if (!I.insert(r)) { S.undo(); return false; } return true;}{}

}
\end{code}

% EN: We have to check the return value of every function, undo the first action if the second one fails, and return \texttt{true} only if both actions succeed.
\Trans{We have to check the return value of every function, undo the first action if the second one fails, and return \texttt{true} only if both actions succeed.}{}

% EN: So far, so good; we were able to fix the simplest two-stage problem, so the code is error-safe. Now, it is time to up the complexity. Suppose that our storage needs some cleanup to be performed at the end of the transaction, for example, the inserted record is not in its final state until we call the \texttt{Storage::finalize()} method (maybe this is done so \texttt{Storage::undo()} can work, and after the insertion is finalized it can no longer be undone). Note the difference between \texttt{undo()} and \texttt{finalize()}; the former must be called only if we want to roll back the transaction, while the latter must be called if the storage insertion succeeded, regardless of what happens after.
\Trans{So far, so good; we were able to fix the simplest two-stage problem, so the code is error-safe. Now, it is time to up the complexity. Suppose that our storage needs some cleanup to be performed at the end of the transaction, for example, the inserted record is not in its final state until we call the \texttt{Storage::finalize()} method (maybe this is done so \texttt{Storage::undo()} can work, and after the insertion is finalized it can no longer be undone). Note the difference between \texttt{undo()} and \texttt{finalize()}; the former must be called only if we want to roll back the transaction, while the latter must be called if the storage insertion succeeded, regardless of what happens after.}{}

% EN: Our requirements are met with this flow of control:
\Trans{Our requirements are met with this flow of control:}{}

\begin{code}
% EN: // Example 02a:
void Database::insert(const Record& r) {
  S.insert(r);
  try {
    I.insert(r);
   } catch (...) {
    S.undo();
    S.finalize();
    throw;
   }
  S.finalize();
\Trans{// Example 02a: void Database::insert(const Record& r) { S.insert(r); try { I.insert(r); } catch (...) { S.undo(); S.finalize(); throw; } S.finalize();}{}

}
\end{code}

% EN: Or we have something similar in the case of returning error codes (for the rest of this chapter, we will use exceptions in all our examples, but conversion to error codes is not hard).
\Trans{Or we have something similar in the case of returning error codes (for the rest of this chapter, we will use exceptions in all our examples, but conversion to error codes is not hard).}{}

% EN: This is already getting ugly, particularly the part about getting the cleanup code (in our case, \texttt{S.finalize()}) to run in every execution path. It's only going to get worse if we have a more complex sequence of actions that must all be undone unless the entire operation succeeds. Here is the control flow for three actions, each with its own rollback and cleanup:
\Trans{This is already getting ugly, particularly the part about getting the cleanup code (in our case, \texttt{S.finalize()}) to run in every execution path. It's only going to get worse if we have a more complex sequence of actions that must all be undone unless the entire operation succeeds. Here is the control flow for three actions, each with its own rollback and cleanup:}{}

\begin{code}
% EN: if (action1() == SUCCESS) {
  if (action2() == SUCCESS) {
    if (action3() == FAIL) {
      rollback2();
      rollback1();
     }
    cleanup2();
   } else {
    rollback1();
   }
  cleanup1();
\Trans{if (action1() == SUCCESS) { if (action2() == SUCCESS) { if (action3() == FAIL) { rollback2(); rollback1(); } cleanup2(); } else { rollback1(); } cleanup1();}{}

}
\end{code}

% EN: The obvious problem is the explicit tests for success, either as conditionals or as try-catch blocks. The more serious problem is that this way of error handling is not composable. The solution for N+1 actions is not the code for N actions with some bits added to it; no, we have to go deep inside the code and add the right pieces there. But we have already seen the C++ idiom for solving this very problem.
\Trans{The obvious problem is the explicit tests for success, either as conditionals or as try-catch blocks. The more serious problem is that this way of error handling is not composable. The solution for N+1 actions is not the code for N actions with some bits added to it; no, we have to go deep inside the code and add the right pieces there. But we have already seen the C++ idiom for solving this very problem.}{}

\subsection{Resource Acquisition Is Initialization}

% EN: The RAII idiom binds resources to objects. The object is constructed when the resource is acquired, and the resource is deleted when the object is destroyed. In our case, we are interested only in the second half, the destruction. The advantage of the RAII idiom is that the destructors of all local objects must be called when the control reaches the end of the scope, regardless of how it happens (\texttt{return}, \texttt{throw}, \texttt{break}, and so on). Since we struggled with the cleanup, let's hand that off to an RAII object:
\Trans{The RAII idiom binds resources to objects. The object is constructed when the resource is acquired, and the resource is deleted when the object is destroyed. In our case, we are interested only in the second half, the destruction. The advantage of the RAII idiom is that the destructors of all local objects must be called when the control reaches the end of the scope, regardless of how it happens (\texttt{return}, \texttt{throw}, \texttt{break}, and so on). Since we struggled with the cleanup, let's hand that off to an RAII object:}{}

\begin{code}
% EN: // Example 02b:
class StorageFinalizer {
  public:
  StorageFinalizer(Storage& S) : S_(S) { }
  ~StorageFinalizer() { S_.finalize();  }
  private:
  Storage& S_;
 };
void Database::insert(const Record& r) {
  S.insert(r);
  StorageFinalizer SF(S);
  try {
    I.insert(r);
   } catch (...) {
    S.undo();
    throw;
   }
\Trans{// Example 02b: class StorageFinalizer { public: StorageFinalizer(Storage& S) : S_(S) {} ~StorageFinalizer() { S_.finalize(); } private: Storage& S_; }; void Database::insert(const Record& r) { S.insert(r); StorageFinalizer SF(S); try { I.insert(r); } catch (...) { S.undo(); throw; }}{}

}
\end{code}

% EN: The \texttt{StorageFinalizer} object binds to the \texttt{Storage} object when it is constructed and calls \texttt{finalize()} when it is destroyed. Since there is no way to exit the scope in which the \texttt{StorageFinalizer} object is defined without calling its destructor, we do not need to worry about the control flow, at least for the cleanup; it is going to happen. Note that the \texttt{StorageFinalizer} is properly constructed after the storage insertion succeeds; if the first insertion fails, there is nothing to finalize.
\Trans{The \texttt{StorageFinalizer} object binds to the \texttt{Storage} object when it is constructed and calls \texttt{finalize()} when it is destroyed. Since there is no way to exit the scope in which the \texttt{StorageFinalizer} object is defined without calling its destructor, we do not need to worry about the control flow, at least for the cleanup; it is going to happen. Note that the \texttt{StorageFinalizer} is properly constructed after the storage insertion succeeds; if the first insertion fails, there is nothing to finalize.}{}

% EN: This code works, but it looks somehow half-done; we have two actions that are performed at the end of the function, and the first one (cleanup, or \texttt{finalize()}) is automated while the second one (rollback, or \texttt{undo()}) is not. Also, the technique is still not composable; here is the flow of control for three actions:
\Trans{This code works, but it looks somehow half-done; we have two actions that are performed at the end of the function, and the first one (cleanup, or \texttt{finalize()}) is automated while the second one (rollback, or \texttt{undo()}) is not. Also, the technique is still not composable; here is the flow of control for three actions:}{}

\begin{code}
% EN: class Cleanup1() {
  ~Cleanup1() { cleanup1();  }
  ...
 };
class Cleanup2() {
  ~Cleanup2() { cleanup2();  }
  ...
 };
action1();
Cleanup1 c1;
try {
  action2();
  Cleanup2 c2;
  try {
    action3();
   } catch (...) {
    rollback2();
    throw;
   }
 } catch (...) {
  rollback1();
\Trans{class Cleanup1() { ~Cleanup1() { cleanup1(); } ... }; class Cleanup2() { ~Cleanup2() { cleanup2(); } ... }; action1(); Cleanup1 c1; try { action2(); Cleanup2 c2; try { action3(); } catch (...) { rollback2(); throw; } } catch (...) { rollback1();}{}

}
\end{code}

% EN: Again, to add another action, we have to add a try-catch block deep in the code. On the other hand, the cleanup part by itself is perfectly composable. Consider what the previous code looks like if we don't need to do a rollback:
\Trans{Again, to add another action, we have to add a try-catch block deep in the code. On the other hand, the cleanup part by itself is perfectly composable. Consider what the previous code looks like if we don't need to do a rollback:}{}

\begin{code}
% EN: action1();
Cleanup1 c1;
action2();
Cleanup2 c2;
\Trans{action1(); Cleanup1 c1; action2(); Cleanup2 c2;}{}

\end{code}

% EN: If we need one more action, we simply add two lines at the end of the function, and the cleanup happens in the right order. If we could do the same for the rollback, we would be all set.
\Trans{If we need one more action, we simply add two lines at the end of the function, and the cleanup happens in the right order. If we could do the same for the rollback, we would be all set.}{}

% EN: We cannot simply move the call to \texttt{undo()} into a destructor of another object; the destructors are always called, but the rollback happens only in the case of an error. But we can make the destructor call the rollback conditionally:
\Trans{We cannot simply move the call to \texttt{undo()} into a destructor of another object; the destructors are always called, but the rollback happens only in the case of an error. But we can make the destructor call the rollback conditionally:}{}

\begin{code}
% EN: // Example 03:
class StorageGuard {
  public:
  StorageGuard(Storage& S) : S_(S) { }
  ~StorageGuard() {
    if (!commit_) S_.undo();
   }
  void commit() noexcept { commit_ = true;  }
  private:
  Storage& S_;
  bool commit_ = false;
 };
void Database::insert(const Record& r) {
  S.insert(r);
  StorageFinalizer SF(S);
  StorageGuard SG(S);
  I.insert(r);
  SG.commit();
\Trans{// Example 03: class StorageGuard { public: StorageGuard(Storage& S) : S_(S) {} ~StorageGuard() { if (!commit_) S_.undo(); } void commit() noexcept { commit_ = true; } private: Storage& S_; bool commit_ = false; }; void Database::insert(const Record& r) { S.insert(r); StorageFinalizer SF(S); StorageGuard SG(S); I.insert(r); SG.commit();}{}

}
\end{code}

% EN: Examine the code now; if the storage insertion fails, the exception is thrown and the database is unchanged. If it succeeds, two RAII objects are constructed. The first one will unconditionally call \texttt{S.finalize()} at the end of the scope. The second one will call \texttt{S.undo()} unless we commit the change first by calling the \texttt{commit()} method on the \texttt{StorageGuard} object. That will happen unless the index insertion fails, in which case an exception is thrown, the rest of the code in the scope is bypassed, and the control jumps straight to the end of the scope (the closing \texttt{\}}) where the destructors of all local objects are called. Since we never called \texttt{commit()} in this scenario, the \texttt{StorageGuard} is still active and will undo the insertion. Note also that there are no explicit \texttt{try-catch} blocks at all: the actions that used to be in the \texttt{catch} clause are now done by the destructors. The exception should be, of course, caught eventually (in all examples that accompany this chapter, the exception is caught in \texttt{main()}).
\Trans{Examine the code now; if the storage insertion fails, the exception is thrown and the database is unchanged. If it succeeds, two RAII objects are constructed. The first one will unconditionally call \texttt{S.finalize()} at the end of the scope. The second one will call \texttt{S.undo()} unless we commit the change first by calling the \texttt{commit()} method on the \texttt{StorageGuard} object. That will happen unless the index insertion fails, in which case an exception is thrown, the rest of the code in the scope is bypassed, and the control jumps straight to the end of the scope (the closing \texttt{\}}) where the destructors of all local objects are called. Since we never called \texttt{commit()} in this scenario, the \texttt{StorageGuard} is still active and will undo the insertion. Note also that there are no explicit \texttt{try-catch} blocks at all: the actions that used to be in the \texttt{catch} clause are now done by the destructors. The exception should be, of course, caught eventually (in all examples that accompany this chapter, the exception is caught in \texttt{main()}).}{}

% EN: The destructors of the local objects are called in reverse construction order. This is important; if we have to undo the insertion, this can be done only until the action is finalized, so the rollback has to happen before the cleanup. Therefore, we construct the RAII objects in the correct order---first, the cleanup (to be done last), then the rollback guard (to be done first, if necessary).
\Trans{The destructors of the local objects are called in reverse construction order. This is important; if we have to undo the insertion, this can be done only until the action is finalized, so the rollback has to happen before the cleanup. Therefore, we construct the RAII objects in the correct order---first, the cleanup (to be done last), then the rollback guard (to be done first, if necessary).}{}

% EN: The code now looks very nice, with no try-catch blocks at all. In some ways, it does not look like regular C++. This programming style is called \textbf{declarative programming}; it is a programming paradigm in which the program logic is expressed without explicitly stating the flow of control (the opposite, and the more common in C++, is \textbf{imperative programming}, where the program describes which steps to do and in what order, but not necessarily why). There are declarative programming languages (the prime example is SQL), but C++ is not one of them. Nonetheless, C++ is very good at implementing constructs that allow the creation of higher-order languages on top of C++, and so we have implemented a declarative error-handling language. Our program now says that after the record was inserted into storage, two actions are pending---the cleanup and the rollback.
\Trans{The code now looks very nice, with no try-catch blocks at all. In some ways, it does not look like regular C++. This programming style is called \textbf{declarative programming}; it is a programming paradigm in which the program logic is expressed without explicitly stating the flow of control (the opposite, and the more common in C++, is \textbf{imperative programming}, where the program describes which steps to do and in what order, but not necessarily why). There are declarative programming languages (the prime example is SQL), but C++ is not one of them. Nonetheless, C++ is very good at implementing constructs that allow the creation of higher-order languages on top of C++, and so we have implemented a declarative error-handling language. Our program now says that after the record was inserted into storage, two actions are pending---the cleanup and the rollback.}{}

% EN: The rollback is disarmed if the entire function succeeds. The code looks linear, without the explicit flow of control, in other words, declarative.
\Trans{The rollback is disarmed if the entire function succeeds. The code looks linear, without the explicit flow of control, in other words, declarative.}{}

% EN: Nice as it is, it is also far from perfect. The obvious problem is that we have to write a guard or a finalizer class for every action in our program. The less obvious one is that writing these classes correctly is not easy, and we have not done a particularly good job so far. Take a moment to figure out what is missing before looking at the fixed version here:
\Trans{Nice as it is, it is also far from perfect. The obvious problem is that we have to write a guard or a finalizer class for every action in our program. The less obvious one is that writing these classes correctly is not easy, and we have not done a particularly good job so far. Take a moment to figure out what is missing before looking at the fixed version here:}{}

\begin{code}
% EN: class StorageGuard {
  public:
  StorageGuard(Storage& S) : S_(S), commit_(false) { }
  ~StorageGuard() { if (!commit_) S_.undo();  }
  void commit() noexcept { commit_ = true;  }
  private:
  Storage& S_;
  bool commit_;
  // Important: really bad things happen if
  // this guard is copied!
  StorageGuard(const StorageGuard&) = delete;
  StorageGuard& operator=(const StorageGuard&) = delete;
 };
void Database::insert(const Record& r) {
  S.insert(r);
  StorageFinalizer SF(S);
  StorageGuard SG(S);
  I.insert(r);
  SG.commit();
\Trans{class StorageGuard { public: StorageGuard(Storage& S) : S_(S), commit_(false) {} ~StorageGuard() { if (!commit_) S_.undo(); } void commit() noexcept { commit_ = true; } private: Storage& S_; bool commit_; // Important: really bad things happen if // this guard is copied! StorageGuard(const StorageGuard&) = delete; StorageGuard& operator=(const StorageGuard&) = delete; }; void Database::insert(const Record& r) { S.insert(r); StorageFinalizer SF(S); StorageGuard SG(S); I.insert(r); SG.commit();}{}

}
\end{code}

% EN: What we need is a general framework that lets us schedule an arbitrary action to be executed at the end of the scope, unconditionally or conditionally. The next section presents the pattern that provides such a framework, ScopeGuard.
\Trans{What we need is a general framework that lets us schedule an arbitrary action to be executed at the end of the scope, unconditionally or conditionally. The next section presents the pattern that provides such a framework, ScopeGuard.}{}

\section{The ScopeGuard pattern}

% EN: In this section, we learn how to write the on-exit action RAII classes such as the ones we implemented in the previous section, but without all the boilerplate code. This can be done in C++03 but is much improved in C++14, and again in C++17.
\Trans{In this section, we learn how to write the on-exit action RAII classes such as the ones we implemented in the previous section, but without all the boilerplate code. This can be done in C++03 but is much improved in C++14, and again in C++17.}{}

\subsection{ScopeGuard basics}

% EN: Let's start with the more difficult problem---how to implement a generic rollback class, a generic version of \texttt{StorageGuard} from the last section. The only difference between that and the cleanup class is that the cleanup is always active, but the rollback is canceled after the action is committed. If we have the conditional rollback version, we can always take out the condition check, and we get the cleanup version, so let's not worry about that for now.
\Trans{Let's start with the more difficult problem---how to implement a generic rollback class, a generic version of \texttt{StorageGuard} from the last section. The only difference between that and the cleanup class is that the cleanup is always active, but the rollback is canceled after the action is committed. If we have the conditional rollback version, we can always take out the condition check, and we get the cleanup version, so let's not worry about that for now.}{}

% EN: In our example, the rollback is a call to the \texttt{S.undo()} method. To simplify the example, let's start with a rollback that calls a regular function, not a member function:
\Trans{In our example, the rollback is a call to the \texttt{S.undo()} method. To simplify the example, let's start with a rollback that calls a regular function, not a member function:}{}

\begin{code}
% EN: void undo(Storage& S) { S.undo(); }
\Trans{void undo(Storage& S) { S.undo(); }}{}
\end{code}

% EN: Once the implementation is finished, the program should look something like this:
\Trans{Once the implementation is finished, the program should look something like this:}{}

\begin{code}
{
% EN:   S.insert(r);
  ScopeGuard SG(undo, S);    // Approximate desired syntax
  ...
  SG.commit();            // Disarm the scope guard
\Trans{S.insert(r); ScopeGuard SG(undo, S); // Approximate desired syntax ... SG.commit(); // Disarm the scope guard}{}

}
\end{code}

% EN: This code tells us (in a declarative manner!) that if the insert action succeeded, we schedule the rollback to be done upon exiting the scope. The rollback will call the \texttt{undo()} function with the argument \texttt{S}, which in turn will undo the insertion. If we made it to the end of the function, we disarm the guard and disable the rollback call, which commits the insertion and makes it permanent.
\Trans{This code tells us (in a declarative manner!) that if the insert action succeeded, we schedule the rollback to be done upon exiting the scope. The rollback will call the \texttt{undo()} function with the argument \texttt{S}, which in turn will undo the insertion. If we made it to the end of the function, we disarm the guard and disable the rollback call, which commits the insertion and makes it permanent.}{}

% EN: A much more general and reusable solution was proposed by Andrei Alexandrescu in the \emph{Dr. Dobbs} article in 2000 (http://www.drdobbs.com/cpp/generic-change-the-way-you-write-excepti/184403758?). Let's look at the implementation and analyze it:
\Trans{A much more general and reusable solution was proposed by Andrei Alexandrescu in the \emph{Dr. Dobbs} article in 2000 (http://www.drdobbs.com/cpp/generic-change-the-way-you-write-excepti/184403758?). Let's look at the implementation and analyze it:}{}

\begin{code}
% EN: // Example 04
class ScopeGuardImplBase {
  public:
  ScopeGuardImplBase() = default;
  void commit() const noexcept { commit_ = true;  }
  protected:
  ScopeGuardImplBase(const ScopeGuardImplBase& other) :
    commit_(other.commit_) { other.commit();  }
  ~ScopeGuardImplBase() { }
  mutable bool commit_ = false;
 };
template <typename Func, typename Arg>
class ScopeGuardImpl : public ScopeGuardImplBase {
  public:
  ScopeGuardImpl(const Func& func, Arg& arg) :
    func_(func), arg_(arg) { }
  ~ScopeGuardImpl() { if (!commit_) func_(arg_);  }
  private:
  const Func& func_;
  Arg& arg_;
 };
template <typename Func, typename Arg>
ScopeGuardImpl<Func, Arg>
MakeGuard(const Func& func, Arg& arg) {
  return ScopeGuardImpl<Func, Arg>(func, arg);
\Trans{// Example 04 class ScopeGuardImplBase { public: ScopeGuardImplBase() = default; void commit() const noexcept { commit_ = true; } protected: ScopeGuardImplBase(const ScopeGuardImplBase& other) : commit_(other.commit_) { other.commit(); } ~ScopeGuardImplBase() {} mutable bool commit_ = false; }; template <typename Func, typename Arg> class ScopeGuardImpl : public ScopeGuardImplBase { public: ScopeGuardImpl(const Func& func, Arg& arg) : func_(func), arg_(arg) {} ~ScopeGuardImpl() { if (!commit_) func_(arg_); } private: const Func& func_; Arg& arg_; }; template <typename Func, typename Arg> ScopeGuardImpl<Func, Arg> MakeGuard(const Func& func, Arg& arg) { return ScopeGuardImpl<Func, Arg>(func, arg);}{}

}
\end{code}

% EN: From the top, we have the base class for all the scope guards, \texttt{ScopeGuardImplBase}. The base class holds the commit flag and the code to manipulate it; the constructor initially creates the guard in the armed state, so the deferred action will happen in the destructor. The call to \texttt{commit()} will prevent that from happening and make the destructor do nothing. Finally, there is a copy constructor that creates a new guard in the same state as the original but then disarms the original guard. This is so the rollback does not happen twice, from the destructors of both objects. The object is copyable but not assignable. We use C++03 features for everything here, including the disabled assignment operator. This implementation is fundamentally C++03; the few C++11 twists are just icing on the cake (that is about to change in the next section).
\Trans{From the top, we have the base class for all the scope guards, \texttt{ScopeGuardImplBase}. The base class holds the commit flag and the code to manipulate it; the constructor initially creates the guard in the armed state, so the deferred action will happen in the destructor. The call to \texttt{commit()} will prevent that from happening and make the destructor do nothing. Finally, there is a copy constructor that creates a new guard in the same state as the original but then disarms the original guard. This is so the rollback does not happen twice, from the destructors of both objects. The object is copyable but not assignable. We use C++03 features for everything here, including the disabled assignment operator. This implementation is fundamentally C++03; the few C++11 twists are just icing on the cake (that is about to change in the next section).}{}

% EN: There are several details of the implementation of \texttt{ScopeGuardImplBase} that may seem odd and need elaboration. First of all, the destructor is not virtual; this is not a typo or a bug, this is intentional, as we will see later. Second, the \texttt{commit\_} flag is declared \texttt{mutable}. This is, of course, so that it can be changed by the \texttt{commit()} method, which we declared \texttt{const}. So, why is \texttt{commit()} declared \texttt{const}? One reason is so we can call it on the source object from the copy constructor, to transfer the responsibility for the rollback from the other object to this one. In this sense, the copy constructor really does a move and will be officially declared as such later. The second reason for \texttt{const} will become obvious later (it is related to the non-virtual destructor, of all things).
\Trans{There are several details of the implementation of \texttt{ScopeGuardImplBase} that may seem odd and need elaboration. First of all, the destructor is not virtual; this is not a typo or a bug, this is intentional, as we will see later. Second, the \texttt{commit\_} flag is declared \texttt{mutable}. This is, of course, so that it can be changed by the \texttt{commit()} method, which we declared \texttt{const}. So, why is \texttt{commit()} declared \texttt{const}? One reason is so we can call it on the source object from the copy constructor, to transfer the responsibility for the rollback from the other object to this one. In this sense, the copy constructor really does a move and will be officially declared as such later. The second reason for \texttt{const} will become obvious later (it is related to the non-virtual destructor, of all things).}{}

% EN: Let's now turn to the derived class, \texttt{ScopeGuardImpl}. This is a class template with two type parameters---the first is the type of the function or any other callable object we are going to call for the rollback, and the second is the type of the argument. Our rollback function is restricted, for now, to having only one argument. That function is called in the destructor of the \texttt{ScopeGuard} object unless the guard was disarmed by calling \texttt{commit()}.
\Trans{Let's now turn to the derived class, \texttt{ScopeGuardImpl}. This is a class template with two type parameters---the first is the type of the function or any other callable object we are going to call for the rollback, and the second is the type of the argument. Our rollback function is restricted, for now, to having only one argument. That function is called in the destructor of the \texttt{ScopeGuard} object unless the guard was disarmed by calling \texttt{commit()}.}{}

% EN: Finally, we have a factory function template, \texttt{MakeGuard}. This is a very common idiom in C++; if you need to create an instantiation of a class template from the constructor arguments, use a template function that can deduce the type of the parameters and that of the return value from the arguments (in C++17 class templates can do that too, as we will see later).
\Trans{Finally, we have a factory function template, \texttt{MakeGuard}. This is a very common idiom in C++; if you need to create an instantiation of a class template from the constructor arguments, use a template function that can deduce the type of the parameters and that of the return value from the arguments (in C++17 class templates can do that too, as we will see later).}{}

% EN: How is this all used to create a guard object that will call \texttt{undo(S)} for us? Like this:
\Trans{How is this all used to create a guard object that will call \texttt{undo(S)} for us? Like this:}{}

\begin{code}
% EN: void Database::insert(const Record& r) {
  S.insert(r);
  const ScopeGuardImplBase& SG = MakeGuard(undo, S);
  I.insert(r);
  SG.commit();
\Trans{void Database::insert(const Record& r) { S.insert(r); const ScopeGuardImplBase& SG = MakeGuard(undo, S); I.insert(r); SG.commit();}{}

}
\end{code}

% EN: The \texttt{MakeGaurd} function deduces the types of the \texttt{undo()} function and the argument \texttt{S} and returns a \texttt{ScopeGuard} object of the corresponding type. The return is by value, so there is a copy involved (the compiler may choose to elide the copy as an optimization, but is not required to). The returned object is a temporary variable, it has no name and binds it to the base class reference \texttt{SG} (casting from the derived class to the base class is implicit for both pointers and references). The lifetime of the temporary variable is until the closing semicolon of the statement that created it, as everyone knows. But then, what does the \texttt{SG} reference point to after the end of the statement? It has to be bound to something, as references cannot be unbound, they are not like \texttt{NULL} pointers. The truth is, ``everyone'' knows wrong, or rather only mostly right---usually, the temporaries indeed live until the end of the statement. However, binding a temporary to a \texttt{const} reference extends its lifetime to coincide with the lifetime of the reference itself. In other words, the unnamed temporary \texttt{ScopeGuard} object created by \texttt{MakeGuard} will not be destroyed until the \texttt{SG} reference goes out of scope. Const-ness is important here, but do not worry, you can't forget it; the language does not permit binding non-\texttt{const} references to temporary variables, so the compiler will let you know. So this explains the \texttt{commit()} method; it has to be \texttt{const} since we are going to call it on the \texttt{const} reference (and, therefore, the \texttt{commit\_} flag has to be \texttt{mutable}).
\Trans{The \texttt{MakeGaurd} function deduces the types of the \texttt{undo()} function and the argument \texttt{S} and returns a \texttt{ScopeGuard} object of the corresponding type. The return is by value, so there is a copy involved (the compiler may choose to elide the copy as an optimization, but is not required to). The returned object is a temporary variable, it has no name and binds it to the base class reference \texttt{SG} (casting from the derived class to the base class is implicit for both pointers and references). The lifetime of the temporary variable is until the closing semicolon of the statement that created it, as everyone knows. But then, what does the \texttt{SG} reference point to after the end of the statement? It has to be bound to something, as references cannot be unbound, they are not like \texttt{NULL} pointers. The truth is, ``everyone'' knows wrong, or rather only mostly right---usually, the temporaries indeed live until the end of the statement. However, binding a temporary to a \texttt{const} reference extends its lifetime to coincide with the lifetime of the reference itself. In other words, the unnamed temporary \texttt{ScopeGuard} object created by \texttt{MakeGuard} will not be destroyed until the \texttt{SG} reference goes out of scope. Const-ness is important here, but do not worry, you can't forget it; the language does not permit binding non-\texttt{const} references to temporary variables, so the compiler will let you know. So this explains the \texttt{commit()} method; it has to be \texttt{const} since we are going to call it on the \texttt{const} reference (and, therefore, the \texttt{commit\_} flag has to be \texttt{mutable}).}{}

% EN: But what about the destructor? At the end of the scope, the destructor of the \texttt{ScopeGuardImplBase} class will be called, since that is the type of reference that goes out of scope. The base class destructor itself does nothing, it's the derived class that has the destructor we want. A polymorphic class with a virtual destructor would have served us right, but we did not take that route. Instead, we availed ourselves of yet another special rule in the C++ standard concerning the \texttt{const} references and temporary variables---not only is the lifetime of the temporary variable extended but also the destructor of the derived class, that is the actual class that was constructed, is going to be called at the end of the scope.
\Trans{But what about the destructor? At the end of the scope, the destructor of the \texttt{ScopeGuardImplBase} class will be called, since that is the type of reference that goes out of scope. The base class destructor itself does nothing, it's the derived class that has the destructor we want. A polymorphic class with a virtual destructor would have served us right, but we did not take that route. Instead, we availed ourselves of yet another special rule in the C++ standard concerning the \texttt{const} references and temporary variables---not only is the lifetime of the temporary variable extended but also the destructor of the derived class, that is the actual class that was constructed, is going to be called at the end of the scope.}{}

% EN: Note that this rule applies only to destructors; you still cannot call derived class methods on the base class \texttt{SG} reference. Also, the lifetime extension works only when the temporary variable is directly bound to a \texttt{const} reference. It does not work if, for example, we initialize another \texttt{const} reference from the first one. This is why we had to return the \texttt{ScopeGuard} object from the \texttt{MakeGuard} function by value; if we tried to return it by reference, the temporary would be bound to that reference, which is going to go away at the end of the statement. The second reference, \texttt{SG}, initialized from the first one, would not have extended the lifetime of the object.
\Trans{Note that this rule applies only to destructors; you still cannot call derived class methods on the base class \texttt{SG} reference. Also, the lifetime extension works only when the temporary variable is directly bound to a \texttt{const} reference. It does not work if, for example, we initialize another \texttt{const} reference from the first one. This is why we had to return the \texttt{ScopeGuard} object from the \texttt{MakeGuard} function by value; if we tried to return it by reference, the temporary would be bound to that reference, which is going to go away at the end of the statement. The second reference, \texttt{SG}, initialized from the first one, would not have extended the lifetime of the object.}{}

% EN: The function implementation we just saw comes very close to the original goal, it is just a bit more verbose (and it mentions \texttt{ScopeGuardImplBase} instead of the promised \texttt{ScopeGuard}). Fear not, for the last step is merely syntactic sugar:
\Trans{The function implementation we just saw comes very close to the original goal, it is just a bit more verbose (and it mentions \texttt{ScopeGuardImplBase} instead of the promised \texttt{ScopeGuard}). Fear not, for the last step is merely syntactic sugar:}{}

\begin{code}
% EN: using ScopeGuard = const ScopeGuardImplBase&;
\Trans{using ScopeGuard = const ScopeGuardImplBase&;}{}
\end{code}

% EN: Now, we can write this:
\Trans{Now, we can write this:}{}

\begin{code}
% EN: // Example 04a
void Database::insert(const Record& r) {
  S.insert(r);
  ScopeGuard SG = MakeGuard(undo, S);
  I.insert(r);
  SG.commit();
\Trans{// Example 04a void Database::insert(const Record& r) { S.insert(r); ScopeGuard SG = MakeGuard(undo, S); I.insert(r); SG.commit();}{}

}
\end{code}

% EN: This is as far as we are going to get with the language tools we've used so far. Ideally, the desired syntax would be as follows (and we are not very far off):
\Trans{This is as far as we are going to get with the language tools we've used so far. Ideally, the desired syntax would be as follows (and we are not very far off):}{}

\begin{code}
% EN: ScopeGuard SG(undo, S);
\Trans{ScopeGuard SG(undo, S);}{}
\end{code}

% EN: We can tidy up our \texttt{ScopeGuard} a bit by making use of the C++11 features. First of all, we can properly disable the assignment operator. Second, we can stop pretending that our copy constructor is anything other than a move constructor:
\Trans{We can tidy up our \texttt{ScopeGuard} a bit by making use of the C++11 features. First of all, we can properly disable the assignment operator. Second, we can stop pretending that our copy constructor is anything other than a move constructor:}{}

\begin{code}
% EN: // Example 05
class ScopeGuardImplBase {
  public:
  ScopeGuardImplBase() = default;
  void commit() const noexcept { commit_ = true;  }
  protected:
  ScopeGuardImplBase(ScopeGuardImplBase&& other) :
    commit_(other.commit_) { other.commit();  }
  ~ScopeGuardImplBase() { }
  mutable bool commit_ = false;
  private:
  ScopeGuardImplBase& operator=(const ScopeGuardImplBase&)
    = delete;
 };
using ScopeGuard = const ScopeGuardImplBase&;
template <typename Func, typename Arg>
class ScopeGuardImpl : public ScopeGuardImplBase {
  public:
  ScopeGuardImpl(const Func& func, Arg& arg) :
    func_(func), arg_(arg) { }
  ~ScopeGuardImpl() { if (!commit_) func_(arg_);  }
  ScopeGuardImpl(ScopeGuardImpl&& other) :
    ScopeGuardImplBase(std::move(other)),
    func_(other.func_),
    arg_(other.arg_) { }
  private:
  const Func& func_;
  Arg& arg_;
 };
template <typename Func, typename Arg>
ScopeGuardImpl<Func, Arg>
MakeGuard(const Func& func, Arg& arg) {
  return ScopeGuardImpl<Func, Arg>(func, arg);
\Trans{// Example 05 class ScopeGuardImplBase { public: ScopeGuardImplBase() = default; void commit() const noexcept { commit_ = true; } protected: ScopeGuardImplBase(ScopeGuardImplBase&& other) : commit_(other.commit_) { other.commit(); } ~ScopeGuardImplBase() {} mutable bool commit_ = false; private: ScopeGuardImplBase& operator=(const ScopeGuardImplBase&) = delete; }; using ScopeGuard = const ScopeGuardImplBase&; template <typename Func, typename Arg> class ScopeGuardImpl : public ScopeGuardImplBase { public: ScopeGuardImpl(const Func& func, Arg& arg) : func_(func), arg_(arg) {} ~ScopeGuardImpl() { if (!commit_) func_(arg_); } ScopeGuardImpl(ScopeGuardImpl&& other) : ScopeGuardImplBase(std::move(other)), func_(other.func_), arg_(other.arg_) {} private: const Func& func_; Arg& arg_; }; template <typename Func, typename Arg> ScopeGuardImpl<Func, Arg> MakeGuard(const Func& func, Arg& arg) { return ScopeGuardImpl<Func, Arg>(func, arg);}{}

}
\end{code}

% EN: Moving to C++14, we can make one more simplification and deduce the return type of the \texttt{MakeGuard} function:
\Trans{Moving to C++14, we can make one more simplification and deduce the return type of the \texttt{MakeGuard} function:}{}

\begin{code}
% EN: // Example 05a
template <typename Func, typename Arg>
auto MakeGuard(const Func& func, Arg& arg) {
  return ScopeGuardImpl<Func, Arg>(func, arg);
\Trans{// Example 05a template <typename Func, typename Arg> auto MakeGuard(const Func& func, Arg& arg) { return ScopeGuardImpl<Func, Arg>(func, arg);}{}

}
\end{code}

% EN: There is still one concession we had to make---we did not really need the \texttt{undo(S)} function what we really wanted was to call \texttt{S.undo()}. This is done just as easily with a member function variant of \texttt{ScopeGuard}. In fact, the only reason we have not done so from the beginning is to make the example easier to follow; the member function pointer syntax is not the most straightforward aspect of C++:
\Trans{There is still one concession we had to make---we did not really need the \texttt{undo(S)} function what we really wanted was to call \texttt{S.undo()}. This is done just as easily with a member function variant of \texttt{ScopeGuard}. In fact, the only reason we have not done so from the beginning is to make the example easier to follow; the member function pointer syntax is not the most straightforward aspect of C++:}{}

\begin{code}
% EN: // Example 06
template <typename MemFunc, typename Obj>
class ScopeGuardImpl : public ScopeGuardImplBase {
  public:
  ScopeGuardImpl(const MemFunc& memfunc, Obj& obj) :
    memfunc_(memfunc), obj_(obj) { }
  ~ScopeGuardImpl() { if (!commit_) (obj_.*memfunc_)();  }
  ScopeGuardImpl(ScopeGuardImpl&& other) :
    ScopeGuardImplBase(std::move(other)),
    memfunc_(other.memfunc_),
    obj_(other.obj_) { }
  private:
  const MemFunc& memfunc_; Obj& obj_;
 };
template <typename MemFunc, typename Obj>
auto MakeGuard(const MemFunc& memfunc, Obj& obj) {// C++14
  return ScopeGuardImpl<MemFunc, Obj>(memfunc, obj);
\Trans{// Example 06 template <typename MemFunc, typename Obj> class ScopeGuardImpl : public ScopeGuardImplBase { public: ScopeGuardImpl(const MemFunc& memfunc, Obj& obj) : memfunc_(memfunc), obj_(obj) {} ~ScopeGuardImpl() { if (!commit_) (obj_.*memfunc_)(); } ScopeGuardImpl(ScopeGuardImpl&& other) : ScopeGuardImplBase(std::move(other)), memfunc_(other.memfunc_), obj_(other.obj_) {} private: const MemFunc& memfunc_; Obj& obj_; }; template <typename MemFunc, typename Obj> auto MakeGuard(const MemFunc& memfunc, Obj& obj) {// C++14 return ScopeGuardImpl<MemFunc, Obj>(memfunc, obj);}{}

}
\end{code}

% EN: Of course, if both versions of the ScopeGuard template are used in the same program, we have to rename one of them. Also, our function guard is limited to calling functions with only one argument, while our member function guard can call only member functions with no arguments. In C++03, this problem is solved in a tedious but reliable way---we have to create versions of the implementation, \texttt{ScopeGuardImpl0}, \texttt{ScopeGuardImp1}, \texttt{ScopeGuardImpl2}, and so on, for functions with zero, one, two, and so on arguments. We then create \texttt{ScopeObjGuardImpl0}, \texttt{ScopeObjGuardImpl1}, and so on, for member functions with zero, one, two, and so on arguments. If we do not create enough, the compiler will let us know. The base class remains the same for all these variants of the derived class, and so is the \texttt{ScopeGuard} \texttt{typedef}.
\Trans{Of course, if both versions of the ScopeGuard template are used in the same program, we have to rename one of them. Also, our function guard is limited to calling functions with only one argument, while our member function guard can call only member functions with no arguments. In C++03, this problem is solved in a tedious but reliable way---we have to create versions of the implementation, \texttt{ScopeGuardImpl0}, \texttt{ScopeGuardImp1}, \texttt{ScopeGuardImpl2}, and so on, for functions with zero, one, two, and so on arguments. We then create \texttt{ScopeObjGuardImpl0}, \texttt{ScopeObjGuardImpl1}, and so on, for member functions with zero, one, two, and so on arguments. If we do not create enough, the compiler will let us know. The base class remains the same for all these variants of the derived class, and so is the \texttt{ScopeGuard} \texttt{typedef}.}{}

% EN: In C++11, we have variadic templates that are designed to address this exact problem, but we're not going to see such an implementation here. There is no reason for it; we can do so much better than that, as you are about to see.
\Trans{In C++11, we have variadic templates that are designed to address this exact problem, but we're not going to see such an implementation here. There is no reason for it; we can do so much better than that, as you are about to see.}{}

\subsection{Generic ScopeGuard}

% EN: We are now firmly in C++11 territory, nothing that you are about to see has a C++03 equivalent of any practical worth.
\Trans{We are now firmly in C++11 territory, nothing that you are about to see has a C++03 equivalent of any practical worth.}{}

% EN: Our \texttt{ScopeGuard} has, so far, allowed us to all arbitrary functions as rollback for any action. Just like the hand-crafted guard objects, the scope guards are composable and guarantee exception safety. But our implementation, so far, is somewhat limited in what exactly we can call to implement the rollback; it has to be a function or a member function. While this seems to cover a lot, we may want to call, for example, two functions to do a single rollback. We could, of course, write a wrapper function for that, but that sets us back on the path toward single-purpose hand-crafted rollback objects.
\Trans{Our \texttt{ScopeGuard} has, so far, allowed us to all arbitrary functions as rollback for any action. Just like the hand-crafted guard objects, the scope guards are composable and guarantee exception safety. But our implementation, so far, is somewhat limited in what exactly we can call to implement the rollback; it has to be a function or a member function. While this seems to cover a lot, we may want to call, for example, two functions to do a single rollback. We could, of course, write a wrapper function for that, but that sets us back on the path toward single-purpose hand-crafted rollback objects.}{}

% EN: There is, in truth, another problem with our implementation. We decided to capture the function argument by reference:
\Trans{There is, in truth, another problem with our implementation. We decided to capture the function argument by reference:}{}

\begin{code}
% EN: ScopeGuardImpl(const Func& func, Arg& arg);
\Trans{ScopeGuardImpl(const Func& func, Arg& arg);}{}
\end{code}

% EN: This mostly works, unless the argument is a constant or a temporary variable; then, our code will not compile.
\Trans{This mostly works, unless the argument is a constant or a temporary variable; then, our code will not compile.}{}

% EN: C++11 gives us another way to create arbitrary callable objects: lambda expressions. Lambdas are actually classes, but they behave like functions in that they can be called with parentheses. They can take arguments, but they can also capture any arguments from the containing scope, and that often obviates the need to pass arguments to the function call itself. We can also write arbitrary code and package it in a lambda expression. This sounds ideal for the scope guard; we could just write something that says, \emph{at the end of the scope run} \emph{this code}.
\Trans{C++11 gives us another way to create arbitrary callable objects: lambda expressions. Lambdas are actually classes, but they behave like functions in that they can be called with parentheses. They can take arguments, but they can also capture any arguments from the containing scope, and that often obviates the need to pass arguments to the function call itself. We can also write arbitrary code and package it in a lambda expression. This sounds ideal for the scope guard; we could just write something that says, \emph{at the end of the scope run} \emph{this code}.}{}

% EN: Let's see what a lambda expression ScopeGuard looks like:
\Trans{Let's see what a lambda expression ScopeGuard looks like:}{}

\begin{code}
% EN: // Example 07
class ScopeGuardBase {
  public:
  ScopeGuardBase() = default;
  void commit() noexcept { commit_ = true;  }
  protected:
  ScopeGuardBase(ScopeGuardBase&& other) noexcept :
    commit_(other.commit_) { other.commit();  }
  ~ScopeGuardBase() = default;
  bool commit_ = false;
  private:
  ScopeGuardBase& operator=(const ScopeGuardBase&)
    = delete;
 };
template <typename Func>
class ScopeGuard : public ScopeGuardBase {
  public:
  ScopeGuard(Func&& func) : func_(std::move(func)) { }
  ScopeGuard(const Func& func) : func_(func) { }
  ~ScopeGuard() { if (!commit_) func_();  }
  ScopeGuard(ScopeGuard&& other) = default;
  private:
  Func func_;
 };
template <typename Func>
ScopeGuard<Func> MakeGuard(Func&& func) {
  return ScopeGuard<Func>(std::forward<Func>(func));
\Trans{// Example 07 class ScopeGuardBase { public: ScopeGuardBase() = default; void commit() noexcept { commit_ = true; } protected: ScopeGuardBase(ScopeGuardBase&& other) noexcept : commit_(other.commit_) { other.commit(); } ~ScopeGuardBase() = default; bool commit_ = false; private: ScopeGuardBase& operator=(const ScopeGuardBase&) = delete; }; template <typename Func> class ScopeGuard : public ScopeGuardBase { public: ScopeGuard(Func&& func) : func_(std::move(func)) {} ScopeGuard(const Func& func) : func_(func) {} ~ScopeGuard() { if (!commit_) func_(); } ScopeGuard(ScopeGuard&& other) = default; private: Func func_; }; template <typename Func> ScopeGuard<Func> MakeGuard(Func&& func) { return ScopeGuard<Func>(std::forward<Func>(func));}{}

}
\end{code}

% EN: The base class is essentially the same as before, except we are not going to use the \texttt{const} reference trick anymore, and so the \texttt{Impl} suffix is gone; what you see is not an implementation aid but the base of the guard class itself; it contains the reusable code dealing with the \texttt{commit\_} flag. Since we do not use the \texttt{const} reference, we can stop pretending that the \texttt{commit()} method is const, and drop the \texttt{mutable} declaration from \texttt{commit\_}.
\Trans{The base class is essentially the same as before, except we are not going to use the \texttt{const} reference trick anymore, and so the \texttt{Impl} suffix is gone; what you see is not an implementation aid but the base of the guard class itself; it contains the reusable code dealing with the \texttt{commit\_} flag. Since we do not use the \texttt{const} reference, we can stop pretending that the \texttt{commit()} method is const, and drop the \texttt{mutable} declaration from \texttt{commit\_}.}{}

% EN: The derived class, on the other hand, is much different. First of all, there is only one class for all types of rollback, and the argument type parameter is gone; instead, we have a functional object that is going to be a lambda, and it'll contain all the arguments that it needs. The destructor is the same as before (except for the missing argument to the callable \texttt{func\_}), and so is the move constructor. But the primary constructor of the object is quite different; the callable object is stored by value, and initialized from either a \texttt{const} reference or an r-value reference, with a suitable overload selected automatically by the compiler.
\Trans{The derived class, on the other hand, is much different. First of all, there is only one class for all types of rollback, and the argument type parameter is gone; instead, we have a functional object that is going to be a lambda, and it'll contain all the arguments that it needs. The destructor is the same as before (except for the missing argument to the callable \texttt{func\_}), and so is the move constructor. But the primary constructor of the object is quite different; the callable object is stored by value, and initialized from either a \texttt{const} reference or an r-value reference, with a suitable overload selected automatically by the compiler.}{}

% EN: The \texttt{MakeGuard} function is largely unchanged, and we do not need two of them; we can use perfect forwarding (\texttt{std::forward}) to forward the argument of any type to one of the constructors of \texttt{ScopeGuard}.
\Trans{The \texttt{MakeGuard} function is largely unchanged, and we do not need two of them; we can use perfect forwarding (\texttt{std::forward}) to forward the argument of any type to one of the constructors of \texttt{ScopeGuard}.}{}

% EN: Here is how this ScopeGuard is used:
\Trans{Here is how this ScopeGuard is used:}{}

\begin{code}
% EN: void Database::insert(const Record& r) {
  S.insert(r);
  auto SG = MakeGuard([&] { S.undo();  });
  I.insert(r);
  SG.commit();
\Trans{void Database::insert(const Record& r) { S.insert(r); auto SG = MakeGuard([&] { S.undo(); }); I.insert(r); SG.commit();}{}

}
\end{code}

% EN: The punctuation-rich construct that is used as an argument to \texttt{MakeGuard} is the lambda expression. It creates a callable object, and calling this object will run the code in the body of the lambda, in our case \texttt{S.undo()}. There is no \texttt{S} variable declared in the lambda object itself, so it has to be captured from the containing scope. All captures are done by reference (\texttt{{[}\&{]}}). Finally, the object is called with no arguments; the parentheses can be omitted, although \texttt{MakeGuard({[}\&{]}()\ \{\ S.undo();\ \});} is also valid. The function does not return anything, that is, the return type is \texttt{void}; it does not have to be explicitly declared. Note that, so far, we have used C++11 lambdas and have not taken advantage of the more powerful C++14 lambdas. This will usually be the case with ScopeGuard, although in practice, you would probably use C++14 just for auto-deduced return types, if nothing else.
\Trans{The punctuation-rich construct that is used as an argument to \texttt{MakeGuard} is the lambda expression. It creates a callable object, and calling this object will run the code in the body of the lambda, in our case \texttt{S.undo()}. There is no \texttt{S} variable declared in the lambda object itself, so it has to be captured from the containing scope. All captures are done by reference (\texttt{{[}\&{]}}). Finally, the object is called with no arguments; the parentheses can be omitted, although \texttt{MakeGuard({[}\&{]}()\ \{\ S.undo();\ \});} is also valid. The function does not return anything, that is, the return type is \texttt{void}; it does not have to be explicitly declared. Note that, so far, we have used C++11 lambdas and have not taken advantage of the more powerful C++14 lambdas. This will usually be the case with ScopeGuard, although in practice, you would probably use C++14 just for auto-deduced return types, if nothing else.}{}

% EN: We have, until now, intentionally set aside the matter of the regular cleanup and focused on error handling and the rollback. Now that we have a decent working ScopeGuard, we can tie up loose ends quite easily:
\Trans{We have, until now, intentionally set aside the matter of the regular cleanup and focused on error handling and the rollback. Now that we have a decent working ScopeGuard, we can tie up loose ends quite easily:}{}

\begin{code}
% EN: // Example 07a
void Database::insert(const Record& r) {
  S.insert(r);
  auto SF = MakeGuard([&] { S.finalize();  });
  auto SG = MakeGuard([&] { S.undo();  });
  I.insert(r);
  SG.commit();
\Trans{// Example 07a void Database::insert(const Record& r) { S.insert(r); auto SF = MakeGuard([&] { S.finalize(); }); auto SG = MakeGuard([&] { S.undo(); }); I.insert(r); SG.commit();}{}

}
\end{code}

% EN: As you can see, nothing special needs to be added to our framework to support the cleanup. We simply create another ScopeGuard that we never disarm.
\Trans{As you can see, nothing special needs to be added to our framework to support the cleanup. We simply create another ScopeGuard that we never disarm.}{}

% EN: We should also point out that, in C++17, we no longer need the \texttt{MakeGuard} function since the compiler can deduce the template arguments from the constructor:
\Trans{We should also point out that, in C++17, we no longer need the \texttt{MakeGuard} function since the compiler can deduce the template arguments from the constructor:}{}

\begin{code}
% EN: // Example 07b
void Database::insert(const Record& r) {
  S.insert(r);
  ScopeGuard SF = [&] { S.finalize();  };    // C++17
  ScopeGuard SG = [&] { S.undo();  };
  I.insert(r);
  SG.commit();
\Trans{// Example 07b void Database::insert(const Record& r) { S.insert(r); ScopeGuard SF = [&] { S.finalize(); }; // C++17 ScopeGuard SG = [&] { S.undo(); }; I.insert(r); SG.commit();}{}

}
\end{code}

% EN: As long as we are on the subject of making ScopeGuard use prettier, we should consider some helpful macros. We can easily write a macro for the cleanup guard, the one that is always executed. We would like the resulting syntax to look something like this (if this isn't declarative enough, I don't know what is):
\Trans{As long as we are on the subject of making ScopeGuard use prettier, we should consider some helpful macros. We can easily write a macro for the cleanup guard, the one that is always executed. We would like the resulting syntax to look something like this (if this isn't declarative enough, I don't know what is):}{}

\begin{code}
% EN: ON_SCOPE_EXIT { S.finalize(); };
\Trans{ON_SCOPE_EXIT { S.finalize(); };}{}
\end{code}

% EN: We can, in fact, get that very syntax. First of all, we need to generate a name for the guard, what used to be called \texttt{SF}, and we need it to be something unique. From the cutting edge of modern C++, we are now reaching decades back, to classic C and its preprocessor tricks, to generate a unique name for an anonymous variable:
\Trans{We can, in fact, get that very syntax. First of all, we need to generate a name for the guard, what used to be called \texttt{SF}, and we need it to be something unique. From the cutting edge of modern C++, we are now reaching decades back, to classic C and its preprocessor tricks, to generate a unique name for an anonymous variable:}{}

\begin{code}
% EN: #define CONCAT2(x, y) x##y
#define CONCAT(x, y) CONCAT2(x, y)
#ifdef __COUNTER__
#define ANON_VAR(x) CONCAT(x, __COUNTER__)
#else
#define ANON_VAR(x) CONCAT(x, __LINE__)
#endif
\Trans{#define CONCAT2(x, y) x##y #define CONCAT(x, y) CONCAT2(x, y) #ifdef __COUNTER__ #define ANON_VAR(x) CONCAT(x, __COUNTER__) #else #define ANON_VAR(x) CONCAT(x, __LINE__) #endif}{}

\end{code}

% EN: The \texttt{\_\_CONCAT\_\_} macros are how you concatenate two tokens in the preprocessor (and yes, you need two of them, that's the way the preprocessor works). The first token will be a user-specified prefix, and the second one is something unique. Many compilers support a preprocessor variable, \texttt{\_\_COUNTER\_\_}, that is incremented every time it's used, so it's never the same. However, it is not in the standard. If \texttt{\_\_COUNTER\_\_} is not available, we have to use the line number \texttt{\_\_LINE\_\_} as a unique identifier. Of course, it is only unique if we don't put two guards on the same line, so don't.
\Trans{The \texttt{\_\_CONCAT\_\_} macros are how you concatenate two tokens in the preprocessor (and yes, you need two of them, that's the way the preprocessor works). The first token will be a user-specified prefix, and the second one is something unique. Many compilers support a preprocessor variable, \texttt{\_\_COUNTER\_\_}, that is incremented every time it's used, so it's never the same. However, it is not in the standard. If \texttt{\_\_COUNTER\_\_} is not available, we have to use the line number \texttt{\_\_LINE\_\_} as a unique identifier. Of course, it is only unique if we don't put two guards on the same line, so don't.}{}

% EN: Now that we have a way to generate an anonymous variable name, we can implement the \texttt{ON\_SCOPE\_EXIT} macro. It would be trivial to implement one where the code is passed as a macro argument, but it would not give us the syntax we want; the argument has to be in parentheses, so at best we could get \texttt{ON\_SCOPE\_EXIT(S.finalize();)}. Also, commas in the code confuse the preprocessor, since it interprets them as a separator between the macro arguments. If you look carefully at the syntax we requested, \texttt{ON\_SCOPE\_EXIT\ \{\ S.finalize();\ \};}, you will realize that this macro has no arguments at all, and the body of the lambda expression is just typed after the no-argument macro. The macro expansion, therefore, ends on something that can be followed by an opening curly brace. Here is how this is done:
\Trans{Now that we have a way to generate an anonymous variable name, we can implement the \texttt{ON\_SCOPE\_EXIT} macro. It would be trivial to implement one where the code is passed as a macro argument, but it would not give us the syntax we want; the argument has to be in parentheses, so at best we could get \texttt{ON\_SCOPE\_EXIT(S.finalize();)}. Also, commas in the code confuse the preprocessor, since it interprets them as a separator between the macro arguments. If you look carefully at the syntax we requested, \texttt{ON\_SCOPE\_EXIT\ \{\ S.finalize();\ \};}, you will realize that this macro has no arguments at all, and the body of the lambda expression is just typed after the no-argument macro. The macro expansion, therefore, ends on something that can be followed by an opening curly brace. Here is how this is done:}{}

\begin{code}
% EN: // Example 08
struct ScopeGuardOnExit { };
template <typename Func>
ScopeGuard<Func> operator+(ScopeGuardOnExit, Func&& func) {
  return ScopeGuard<Func>(std::forward<Func>(func));
\Trans{// Example 08 struct ScopeGuardOnExit {}; template <typename Func> ScopeGuard<Func> operator+(ScopeGuardOnExit, Func&& func) { return ScopeGuard<Func>(std::forward<Func>(func));}{}

}
% EN: #define ON_SCOPE_EXIT auto ANON_VAR(SCOPE_EXIT_STATE) = \
  ScopeGuardOnExit() + [&]()
\Trans{#define ON_SCOPE_EXIT auto ANON_VAR(SCOPE_EXIT_STATE) = \ ScopeGuardOnExit() + [&]()}{}

\end{code}

% EN: The macro expansion declares an anonymous variable that starts with \texttt{SCOPE\_EXIT\_STATE}, followed by a unique number, and it ends on the incomplete lambda expression, \texttt{{[}\&{]}()}, that is completed by the code in the curly braces. In order to not have a closing parenthesis of the former \texttt{MakeGuard} function, which the macro cannot generate (the macro is expanded before the lambda body, so it cannot generate any code after that), we have to replace the \texttt{MakeGuard} function (or the \texttt{ScopeGuard} constructor in C++17) with an operator. The choice of the operator does not matter; we use \texttt{+}, but we could use any binary operator. The first argument to the operator is a temporary object of a unique type, it limits the overload resolution only to the \texttt{operator+()} defined previously (the object itself is not used at all, we only need its type). The \texttt{operator+()} itself is exactly what \texttt{MakeGuard} used to be, it deduces the type of the lambda expression and creates the corresponding \texttt{ScopeGuard} object. The only downside of this technique is that the closing semicolon at the end of the \texttt{ON\_SCOPE\_EXIT} statement is required, and should you forget it, the compiler will remind you in the most obscure and opaque way possible.
\Trans{The macro expansion declares an anonymous variable that starts with \texttt{SCOPE\_EXIT\_STATE}, followed by a unique number, and it ends on the incomplete lambda expression, \texttt{{[}\&{]}()}, that is completed by the code in the curly braces. In order to not have a closing parenthesis of the former \texttt{MakeGuard} function, which the macro cannot generate (the macro is expanded before the lambda body, so it cannot generate any code after that), we have to replace the \texttt{MakeGuard} function (or the \texttt{ScopeGuard} constructor in C++17) with an operator. The choice of the operator does not matter; we use \texttt{+}, but we could use any binary operator. The first argument to the operator is a temporary object of a unique type, it limits the overload resolution only to the \texttt{operator+()} defined previously (the object itself is not used at all, we only need its type). The \texttt{operator+()} itself is exactly what \texttt{MakeGuard} used to be, it deduces the type of the lambda expression and creates the corresponding \texttt{ScopeGuard} object. The only downside of this technique is that the closing semicolon at the end of the \texttt{ON\_SCOPE\_EXIT} statement is required, and should you forget it, the compiler will remind you in the most obscure and opaque way possible.}{}

% EN: Our program code can now be further tidied up:
\Trans{Our program code can now be further tidied up:}{}

\begin{code}
% EN: // Example 08
void Database::insert(const Record& r) {
  S.insert(r);
  ON_SCOPE_EXIT { S.finalize();  };
  auto SG = ScopeGuard([&] { S.undo();  });
  I.insert(r);
  SG.commit();
\Trans{// Example 08 void Database::insert(const Record& r) { S.insert(r); ON_SCOPE_EXIT { S.finalize(); }; auto SG = ScopeGuard([&] { S.undo(); }); I.insert(r); SG.commit();}{}

}
\end{code}

% EN: It is tempting to apply the same technique to the second guard. Unfortunately, this is not so simple; we have to know the name of this variable so we can call \texttt{commit()} on it. We can define a similar macro that does not use an anonymous variable but instead takes the user-specified name:
\Trans{It is tempting to apply the same technique to the second guard. Unfortunately, this is not so simple; we have to know the name of this variable so we can call \texttt{commit()} on it. We can define a similar macro that does not use an anonymous variable but instead takes the user-specified name:}{}

\begin{code}
% EN: // Example 08a
#define ON_SCOPE_EXIT_ROLLBACK(NAME) \
  auto NAME = ScopeGuardOnExit() + [&]()
\Trans{// Example 08a #define ON_SCOPE_EXIT_ROLLBACK(NAME) \ auto NAME = ScopeGuardOnExit() + [&]()}{}

\end{code}

% EN: We can use it to complete the conversion of our code:
\Trans{We can use it to complete the conversion of our code:}{}

\begin{code}
% EN: // Example 08a
void Database::insert(const Record& r) {
  S.insert(r);
  ON_SCOPE_EXIT { S.finalize();  };
  ON_SCOPE_EXIT_ROLLBACK(SG){ S.undo();  };
  I.insert(r);
  SG.commit();
\Trans{// Example 08a void Database::insert(const Record& r) { S.insert(r); ON_SCOPE_EXIT { S.finalize(); }; ON_SCOPE_EXIT_ROLLBACK(SG){ S.undo(); }; I.insert(r); SG.commit();}{}

}
\end{code}

% EN: At this point, we should revisit the issue of composability. For three actions, each with its own rollback and cleanup, we now have the following:
\Trans{At this point, we should revisit the issue of composability. For three actions, each with its own rollback and cleanup, we now have the following:}{}

\begin{code}
% EN: action1();
ON_SCOPE_EXIT { cleanup1;  };
ON_SCOPE_EXIT_ROLLBACK(g2){ rollback1();  };
action2();
ON_SCOPE_EXIT { cleanup2;  };
ON_SCOPE_EXIT_ROLLBACK(g4){ rollback2();  };
action3();
g2.commit();
g4.commit();
\Trans{action1(); ON_SCOPE_EXIT { cleanup1; }; ON_SCOPE_EXIT_ROLLBACK(g2){ rollback1(); }; action2(); ON_SCOPE_EXIT { cleanup2; }; ON_SCOPE_EXIT_ROLLBACK(g4){ rollback2(); }; action3(); g2.commit(); g4.commit();}{}

\end{code}

% EN: One can see how this pattern is trivially extended to any number of actions. An observant reader might wonder whether they have noticed a bug in the code, though---should not the rollback guards be dismissed in reverse construction order? This is not a bug, although neither is the reverse order of all \texttt{commit()} calls. The reason is that \texttt{commit()} cannot throw an exception, it was declared \texttt{noexcept}, and indeed its implementation is such that no exception can be thrown. This is vitally important for the ScopeGuard pattern to work; if \texttt{commit()} could throw, then there would be no way to guarantee that all rollback guards are properly disarmed. At the end of the scope, some actions would be rolled back, and others would not, leaving the system in an inconsistent state.
\Trans{One can see how this pattern is trivially extended to any number of actions. An observant reader might wonder whether they have noticed a bug in the code, though---should not the rollback guards be dismissed in reverse construction order? This is not a bug, although neither is the reverse order of all \texttt{commit()} calls. The reason is that \texttt{commit()} cannot throw an exception, it was declared \texttt{noexcept}, and indeed its implementation is such that no exception can be thrown. This is vitally important for the ScopeGuard pattern to work; if \texttt{commit()} could throw, then there would be no way to guarantee that all rollback guards are properly disarmed. At the end of the scope, some actions would be rolled back, and others would not, leaving the system in an inconsistent state.}{}

% EN: While ScopeGuard was primarily designed to make exception-safe code easier to write, the interaction of the ScopeGuard pattern with exceptions is far from trivial, and we should spend more time on it.
\Trans{While ScopeGuard was primarily designed to make exception-safe code easier to write, the interaction of the ScopeGuard pattern with exceptions is far from trivial, and we should spend more time on it.}{}

\section{ScopeGuard and exceptions}

% EN: The ScopeGuard pattern is designed to correctly run various cleanup and rollback operations automatically upon exiting a scope, no matter what caused the exit---normal completion by reaching the end of the scope, an early return, or an exception. This makes writing error-safe code in general, and exception-safe code in particular, much easier; as long as we queued up the right guards after every action, the correct cleanup and error handling will automatically happen. That is, of course, assuming that ScopeGuard itself is functioning correctly in the presence of exceptions. We are going to learn how to make sure it does and how to use it to make the rest of the code error-safe.
\Trans{The ScopeGuard pattern is designed to correctly run various cleanup and rollback operations automatically upon exiting a scope, no matter what caused the exit---normal completion by reaching the end of the scope, an early return, or an exception. This makes writing error-safe code in general, and exception-safe code in particular, much easier; as long as we queued up the right guards after every action, the correct cleanup and error handling will automatically happen. That is, of course, assuming that ScopeGuard itself is functioning correctly in the presence of exceptions. We are going to learn how to make sure it does and how to use it to make the rest of the code error-safe.}{}

\subsection{What must not throw an exception}

% EN: We have already seen that the \texttt{commit()} function that is used to commit an action and disarm the rollback guard must never throw an exception. Fortunately, that is easy to guarantee since all this function does is set a flag. But what happens if the rollback function also fails, and throws an exception?
\Trans{We have already seen that the \texttt{commit()} function that is used to commit an action and disarm the rollback guard must never throw an exception. Fortunately, that is easy to guarantee since all this function does is set a flag. But what happens if the rollback function also fails, and throws an exception?}{}

\begin{code}
% EN: // Example 09
void Database::insert(const Record& r) {
  S.insert(r);
  auto SF = MakeGuard([&] { S.finalize();  });
  auto SG = MakeGuard([&] { S.undo();  });
             // What if undo() can throw?
  I.insert(r);    // Let's say this fails
  SG.commit();    // Commit never happens
 }            // Control jumps here and undo() throws
\Trans{// Example 09 void Database::insert(const Record& r) { S.insert(r); auto SF = MakeGuard([&] { S.finalize(); }); auto SG = MakeGuard([&] { S.undo(); }); // What if undo() can throw? I.insert(r); // Let's say this fails SG.commit(); // Commit never happens } // Control jumps here and undo() throws}{}

\end{code}

% EN: The short answer is \emph{nothing good}. In general, we have a conundrum---we cannot allow the action, in our case the storage insertion, to remain, but we also cannot undo it, since that also fails. Specifically, in C++, two exceptions cannot propagate at the same time. For that reason, the destructors are not allowed to throw exceptions; a destructor may be called when an exception is thrown, and if that destructor also throws, we now have two exceptions propagating at the same time. If this happens, the program immediately terminates. This is not so much a shortcoming of the language as a reflection on the unsolvable nature of the problem in general; we cannot let things stay the way they are, but we also failed in an attempt to change something. There aren't any good options left.
\Trans{The short answer is \emph{nothing good}. In general, we have a conundrum---we cannot allow the action, in our case the storage insertion, to remain, but we also cannot undo it, since that also fails. Specifically, in C++, two exceptions cannot propagate at the same time. For that reason, the destructors are not allowed to throw exceptions; a destructor may be called when an exception is thrown, and if that destructor also throws, we now have two exceptions propagating at the same time. If this happens, the program immediately terminates. This is not so much a shortcoming of the language as a reflection on the unsolvable nature of the problem in general; we cannot let things stay the way they are, but we also failed in an attempt to change something. There aren't any good options left.}{}

% EN: In general, there are three ways a C++ program can handle this situation. The best option is not to fall into this trap---if the rollback cannot throw, none of this will happen. A well-written exception-safe program thus goes to great lengths to provide non-throwing rollback and cleanup. For example, the main action can produce the new data and have it ready, then making the data available to the callers is as simple as swapping a pointer, definitely a non-throwing operation. The rollback involves only swapping the pointer back and maybe deleting something (as we already said, destructors should not throw exceptions; if they do, the program behavior is undefined).
\Trans{In general, there are three ways a C++ program can handle this situation. The best option is not to fall into this trap---if the rollback cannot throw, none of this will happen. A well-written exception-safe program thus goes to great lengths to provide non-throwing rollback and cleanup. For example, the main action can produce the new data and have it ready, then making the data available to the callers is as simple as swapping a pointer, definitely a non-throwing operation. The rollback involves only swapping the pointer back and maybe deleting something (as we already said, destructors should not throw exceptions; if they do, the program behavior is undefined).}{}

% EN: The second option is to suppress the exception in the rollback. We tried to undo the operation, it didn't work, there is nothing else we can do about it, so let's push on. The danger here is that the program may be in an undefined state, and every operation from this point forward may be incorrect. This is, however, the worst-case scenario. In practice, the consequences may be less severe. For example, for our database, we may know that if the rollback fails, there is a chunk of disk space that is claimed by the record but is unreachable from the index. The caller will be correctly informed that the insertion failed, but we have wasted some disk space. This may be preferable to terminating the program outright. If this is what we want, we have to catch any exceptions that might be thrown by the ScopeGuard action:
\Trans{The second option is to suppress the exception in the rollback. We tried to undo the operation, it didn't work, there is nothing else we can do about it, so let's push on. The danger here is that the program may be in an undefined state, and every operation from this point forward may be incorrect. This is, however, the worst-case scenario. In practice, the consequences may be less severe. For example, for our database, we may know that if the rollback fails, there is a chunk of disk space that is claimed by the record but is unreachable from the index. The caller will be correctly informed that the insertion failed, but we have wasted some disk space. This may be preferable to terminating the program outright. If this is what we want, we have to catch any exceptions that might be thrown by the ScopeGuard action:}{}

\begin{code}
% EN: // Example 09a
template <typename Func>
class ScopeGuard : public ScopeGuardBase {
  public:
  ...
  ~ScopeGuard() {
    if (!commit_) try { func_();  } catch (...) { }
   }
  ...
 };
\Trans{// Example 09a template <typename Func> class ScopeGuard : public ScopeGuardBase { public: ... ~ScopeGuard() { if (!commit_) try { func_(); } catch (...) {} } ... };}{}

\end{code}

% EN: The \texttt{catch} clause is empty; we catch everything but do nothing. This implementation is sometimes called a \emph{shielded ScopeGuard}.
\Trans{The \texttt{catch} clause is empty; we catch everything but do nothing. This implementation is sometimes called a \emph{shielded ScopeGuard}.}{}

% EN: The final option is to allow the program to fail. That will happen with no effort on our part if we just let two exceptions happen, but we could also print a message or otherwise signal to the user what is about to happen and why. If we want to insert our own dying action before the program terminates, we have to write code that is very similar to previously:
\Trans{The final option is to allow the program to fail. That will happen with no effort on our part if we just let two exceptions happen, but we could also print a message or otherwise signal to the user what is about to happen and why. If we want to insert our own dying action before the program terminates, we have to write code that is very similar to previously:}{}

\begin{code}
% EN: template <typename Func>
class ScopeGuard : public ScopeGuardBase {
  public:
  ...
  ~ScopeGuard() {
    if (!commit_) try { func_();  } catch (...) {
      std::cout << "Rollback failed" << std::endl;
      throw;    // Rethrow
     }
   }
  ...
 };
\Trans{template <typename Func> class ScopeGuard : public ScopeGuardBase { public: ... ~ScopeGuard() { if (!commit_) try { func_(); } catch (...) { std::cout << "Rollback failed" << std::endl; throw; // Rethrow } } ... };}{}

\end{code}

% EN: The key difference is the \texttt{throw;} statement without any arguments. This re-throws the exception we caught and allows it to continue propagating.
\Trans{The key difference is the \texttt{throw;} statement without any arguments. This re-throws the exception we caught and allows it to continue propagating.}{}

% EN: The difference between the last two code fragments highlights a subtle detail that we have glossed over earlier, but one that will become important later. It is imprecise to say that in C++ the destructors should not throw exceptions. The correct statement is that an exception should not propagate out of the destructor. The destructor can throw anything it wants as long as it also catches it:
\Trans{The difference between the last two code fragments highlights a subtle detail that we have glossed over earlier, but one that will become important later. It is imprecise to say that in C++ the destructors should not throw exceptions. The correct statement is that an exception should not propagate out of the destructor. The destructor can throw anything it wants as long as it also catches it:}{}

\begin{code}
% EN: class LivingDangerously {
  public:
  ~LivingDangerously() {
    try {
      if (cleanup() != SUCCESS) throw 0;
       more_cleanup();
     } catch (...) {
      std::cout << "Cleanup failed, proceeding anyway" <<
      std::endl;
      // No rethrow - this is critical!
     }
   }
 };
\Trans{class LivingDangerously { public: ~LivingDangerously() { try { if (cleanup() != SUCCESS) throw 0; more_cleanup(); } catch (...) { std::cout << "Cleanup failed, proceeding anyway" << std::endl; // No rethrow - this is critical! } } };}{}

\end{code}

% EN: Up until now, we have dealt with exceptions mostly as a nuisance; the program has to remain in a well-defined state if something somewhere throws something, but other than that, we have no use for these exceptions; we just pass them on. On the other hand, our code could work with any type of error handling, be it exceptions or error codes. If we knew for sure that errors are always signaled by exceptions and that any return from a function other than an exception thrown is a success, we could take advantage of that to automate the detection of success or failure and therefore allow a commit or rollback to happen as needed.
\Trans{Up until now, we have dealt with exceptions mostly as a nuisance; the program has to remain in a well-defined state if something somewhere throws something, but other than that, we have no use for these exceptions; we just pass them on. On the other hand, our code could work with any type of error handling, be it exceptions or error codes. If we knew for sure that errors are always signaled by exceptions and that any return from a function other than an exception thrown is a success, we could take advantage of that to automate the detection of success or failure and therefore allow a commit or rollback to happen as needed.}{}

\subsection{Exception-driven ScopeGuard}

% EN: We are now going to assume that if a function returns without throwing an exception, the operation has succeeded. If the function throws, it obviously failed. The objective now is to get rid of the explicit call to \texttt{commit()} and instead detect whether the destructor of ScopeGuard is executed because an exception was thrown, or because the function returned normally.
\Trans{We are now going to assume that if a function returns without throwing an exception, the operation has succeeded. If the function throws, it obviously failed. The objective now is to get rid of the explicit call to \texttt{commit()} and instead detect whether the destructor of ScopeGuard is executed because an exception was thrown, or because the function returned normally.}{}

% EN: There are two parts to this implementation. The first part is specifying when we want the action to be taken. The cleanup guard must be executed regardless of how we exit the scope. The rollback guard is executed only in case of failure. For completeness, we can also have a guard that is executed only if the function has succeeded. The second part is determining what actually happened.
\Trans{There are two parts to this implementation. The first part is specifying when we want the action to be taken. The cleanup guard must be executed regardless of how we exit the scope. The rollback guard is executed only in case of failure. For completeness, we can also have a guard that is executed only if the function has succeeded. The second part is determining what actually happened.}{}

% EN: We will start with the second part. Our ScopeGuard now needs two additional parameters that will tell us whether it should be executed on success and whether it should be executed on failure (both can be enabled at the same time). Only the destructor of \texttt{ScopeGuard} needs to be modified:
\Trans{We will start with the second part. Our ScopeGuard now needs two additional parameters that will tell us whether it should be executed on success and whether it should be executed on failure (both can be enabled at the same time). Only the destructor of \texttt{ScopeGuard} needs to be modified:}{}

\begin{code}
% EN: template <typename Func, bool on_success, bool on_failure>
class ScopeGuard {
  public:
  ...
  ~ScopeGuard() {
    if ((on_success && is_success()) ||
        (on_failure && is_failure())) func_();
   }
  ...
 };
\Trans{template <typename Func, bool on_success, bool on_failure> class ScopeGuard { public: ... ~ScopeGuard() { if ((on_success && is_success()) || (on_failure && is_failure())) func_(); } ... };}{}

\end{code}

% EN: We still need to figure out how to implement the pseudo-functions \texttt{is\_success()} and \texttt{is\_failure()}. Remember that failure means that an exception was thrown. In C++, we have a function for that: \texttt{std::uncaught\_exception()}. It returns true if an exception is currently propagating, and false otherwise. Armed with this knowledge, we can implement our guard:
\Trans{We still need to figure out how to implement the pseudo-functions \texttt{is\_success()} and \texttt{is\_failure()}. Remember that failure means that an exception was thrown. In C++, we have a function for that: \texttt{std::uncaught\_exception()}. It returns true if an exception is currently propagating, and false otherwise. Armed with this knowledge, we can implement our guard:}{}

\begin{code}
% EN: // Example 10
template <typename Func, bool on_success, bool on_failure>
class ScopeGuard {
  public:
  ...
  ~ScopeGuard() {
    if ((on_success && !std::uncaught_exception()) ||
        (on_failure && std::uncaught_exception())) func_();
   }
  ...
 };
\Trans{// Example 10 template <typename Func, bool on_success, bool on_failure> class ScopeGuard { public: ... ~ScopeGuard() { if ((on_success && !std::uncaught_exception()) || (on_failure && std::uncaught_exception())) func_(); } ... };}{}

\end{code}

% EN: Now, back to the first part: \texttt{ScopeGuard} will execute the delayed action if the conditions are right, so how do we tell it what conditions are right? Using the macro approach we developed earlier, we can define three versions of the guard---\texttt{ON\_SCOPE\_EXIT} is always executed, \texttt{ON\_SCOPE\_SUCCESS} is executed only if no exceptions were thrown, and \texttt{ON\_SCOPE\_FAILURE} is executed if an exception was thrown. The latter replaces our \texttt{ON\_SCOPE\_EXIT\_ROLLBACK} macro, only now it too can use an anonymous variable name, since there are no explicit calls to \texttt{commit()}. The three macros are defined in a very similar way, we just need three different unique types instead of one \texttt{ScopeGuardOnExit}, so we can decide which \texttt{operator+()} to call:
\Trans{Now, back to the first part: \texttt{ScopeGuard} will execute the delayed action if the conditions are right, so how do we tell it what conditions are right? Using the macro approach we developed earlier, we can define three versions of the guard---\texttt{ON\_SCOPE\_EXIT} is always executed, \texttt{ON\_SCOPE\_SUCCESS} is executed only if no exceptions were thrown, and \texttt{ON\_SCOPE\_FAILURE} is executed if an exception was thrown. The latter replaces our \texttt{ON\_SCOPE\_EXIT\_ROLLBACK} macro, only now it too can use an anonymous variable name, since there are no explicit calls to \texttt{commit()}. The three macros are defined in a very similar way, we just need three different unique types instead of one \texttt{ScopeGuardOnExit}, so we can decide which \texttt{operator+()} to call:}{}

\begin{code}
% EN: // Example 10
struct ScopeGuardOnExit { };
template <typename Func>
auto operator+(ScopeGuardOnExit, Func&& func) {
  return
    ScopeGuard<Func, true, true>(std::forward<Func>(func));
\Trans{// Example 10 struct ScopeGuardOnExit {}; template <typename Func> auto operator+(ScopeGuardOnExit, Func&& func) { return ScopeGuard<Func, true, true>(std::forward<Func>(func));}{}

}
% EN: #define ON_SCOPE_EXIT auto ANON_VAR(SCOPE_EXIT_STATE) = \
  ScopeGuardOnExit() + [&]()
struct ScopeGuardOnSuccess { };
template <typename Func>
auto operator+(ScopeGuardOnSuccess, Func&& func) {
  return
   ScopeGuard<Func, true, false>(std::forward<Func>(func));
\Trans{#define ON_SCOPE_EXIT auto ANON_VAR(SCOPE_EXIT_STATE) = \ ScopeGuardOnExit() + [&]() struct ScopeGuardOnSuccess {}; template <typename Func> auto operator+(ScopeGuardOnSuccess, Func&& func) { return ScopeGuard<Func, true, false>(std::forward<Func>(func));}{}

}
% EN: #define ON_SCOPE_SUCCESS auto ANON_VAR(SCOPE_EXIT_STATE) =\
  ScopeGuardOnSuccess() + [&]()
struct ScopeGuardOnFailure { };
template <typename Func>
auto operator+(ScopeGuardOnFailure, Func&& func) {
  return
   ScopeGuard<Func, false, true>(std::forward<Func>(func));
\Trans{#define ON_SCOPE_SUCCESS auto ANON_VAR(SCOPE_EXIT_STATE) =\ ScopeGuardOnSuccess() + [&]() struct ScopeGuardOnFailure {}; template <typename Func> auto operator+(ScopeGuardOnFailure, Func&& func) { return ScopeGuard<Func, false, true>(std::forward<Func>(func));}{}

}
% EN: #define ON_SCOPE_FAILURE auto ANON_VAR(SCOPE_EXIT_STATE) =\
  ScopeGuardOnFailure() + [&]()
\Trans{#define ON_SCOPE_FAILURE auto ANON_VAR(SCOPE_EXIT_STATE) =\ ScopeGuardOnFailure() + [&]()}{}

\end{code}

% EN: Each overload of \texttt{operator+()} constructs a \texttt{ScopeGuard} object with different Boolean arguments that control when it does and does not execute. Each macro directs the lambda expression to the desired overload by specifying the type of the first argument to the \texttt{operator+()} using one of the unique tree types we defined just for this purpose: \texttt{ScopeGuardOnExit}, \texttt{ScopeGuardOnSuccess}, and \texttt{ScopeGuardOnFailure}.
\Trans{Each overload of \texttt{operator+()} constructs a \texttt{ScopeGuard} object with different Boolean arguments that control when it does and does not execute. Each macro directs the lambda expression to the desired overload by specifying the type of the first argument to the \texttt{operator+()} using one of the unique tree types we defined just for this purpose: \texttt{ScopeGuardOnExit}, \texttt{ScopeGuardOnSuccess}, and \texttt{ScopeGuardOnFailure}.}{}

% EN: This implementation can pass simple and even fairly elaborate tests and appears to work. Unfortunately, it has a fatal flaw---it does not correctly detect success or failure. To be sure, it works fine if our \texttt{Database::insert()} function was called from the normal control flow, where it may or may not succeed. The problem is we may call \texttt{Database::insert()} from a destructor of some other object, and that object may be used in a scope where an exception is thrown:
\Trans{This implementation can pass simple and even fairly elaborate tests and appears to work. Unfortunately, it has a fatal flaw---it does not correctly detect success or failure. To be sure, it works fine if our \texttt{Database::insert()} function was called from the normal control flow, where it may or may not succeed. The problem is we may call \texttt{Database::insert()} from a destructor of some other object, and that object may be used in a scope where an exception is thrown:}{}

\begin{code}
% EN: class ComplexOperation {
  Database db_;
  public:
  ...
  ~ComplexOperation() {
    try {
      db_.insert(some_record);
     } catch (...) { }    // Shield any exceptions from insert()
   }
 };
\Trans{class ComplexOperation { Database db_; public: ... ~ComplexOperation() { try { db_.insert(some_record); } catch (...) {} // Shield any exceptions from insert() } };}{}

{
% EN:   ComplexOperation OP;
  throw 1;
 }    // OP.~ComplexOperation() runs here
\Trans{ComplexOperation OP; throw 1; } // OP.~ComplexOperation() runs here}{}

\end{code}

% EN: Now, \texttt{db\_.insert()} runs in the presence of an uncaught exception, and so \texttt{std::uncaught\_exception()} will return \texttt{true}. The problem is this is not the exception we were looking for. This exception does not indicate that \texttt{insert()} failed, but it will be treated as such and the database insertion will be undone.
\Trans{Now, \texttt{db\_.insert()} runs in the presence of an uncaught exception, and so \texttt{std::uncaught\_exception()} will return \texttt{true}. The problem is this is not the exception we were looking for. This exception does not indicate that \texttt{insert()} failed, but it will be treated as such and the database insertion will be undone.}{}

% EN: What we really need is to know how many exceptions are currently propagating. This may seem a strange statement since C++ does not allow multiple exceptions to propagate at the same time. However, we have already seen that this is an oversimplification; the second exception can propagate just fine as long as it does not escape the destructor. In the same manner, three or more exceptions can propagate if we have nested destructor calls, we just have to catch them all in time. To solve this problem correctly, we need to know how many exceptions were propagating when the \texttt{Database::insert()} function was called. Then, we can compare it with the number of exceptions propagating at the end of the function, however we got there. If these numbers are the same, \texttt{insert()} did not throw any exceptions, and any preexisting ones are not our concern. If a new exception was added, \texttt{insert()} has failed, and the exit handling must change accordingly.
\Trans{What we really need is to know how many exceptions are currently propagating. This may seem a strange statement since C++ does not allow multiple exceptions to propagate at the same time. However, we have already seen that this is an oversimplification; the second exception can propagate just fine as long as it does not escape the destructor. In the same manner, three or more exceptions can propagate if we have nested destructor calls, we just have to catch them all in time. To solve this problem correctly, we need to know how many exceptions were propagating when the \texttt{Database::insert()} function was called. Then, we can compare it with the number of exceptions propagating at the end of the function, however we got there. If these numbers are the same, \texttt{insert()} did not throw any exceptions, and any preexisting ones are not our concern. If a new exception was added, \texttt{insert()} has failed, and the exit handling must change accordingly.}{}

% EN: C++17 lets us implement this detection; in addition to the previous \texttt{std::uncaught\_exception()}, which is deprecated (and removed in C++20), we now have a new function, \texttt{std::uncaught\_exceptions()}, which returns the number of currently propagating exceptions. We can now implement this \texttt{UncaughtExceptionDetector} to detect new exceptions:
\Trans{C++17 lets us implement this detection; in addition to the previous \texttt{std::uncaught\_exception()}, which is deprecated (and removed in C++20), we now have a new function, \texttt{std::uncaught\_exceptions()}, which returns the number of currently propagating exceptions. We can now implement this \texttt{UncaughtExceptionDetector} to detect new exceptions:}{}

\begin{code}
% EN: // Example 10a
class UncaughtExceptionDetector {
  public:
  UncaughtExceptionDetector() :
    count_(std::uncaught_exceptions()) { }
  operator bool() const noexcept {
    return std::uncaught_exceptions() > count_;
   }
  private:
  const int count_;
 };
\Trans{// Example 10a class UncaughtExceptionDetector { public: UncaughtExceptionDetector() : count_(std::uncaught_exceptions()) {} operator bool() const noexcept { return std::uncaught_exceptions() > count_; } private: const int count_; };}{}

\end{code}

% EN: With this detector, we can finally implement the automatic \texttt{ScopeGuard}:
\Trans{With this detector, we can finally implement the automatic \texttt{ScopeGuard}:}{}

\begin{code}
% EN: // Example 10a
template <typename Func, bool on_success, bool on_failure>
class ScopeGuard {
  public:
  ...
  ~ScopeGuard() {
  if ((on_success && !detector_) ||
      (on_failure && detector_)) func_();
   }
  ...
  private:
  UncaughtExceptionDetector detector_;
  ...
 };
\Trans{// Example 10a template <typename Func, bool on_success, bool on_failure> class ScopeGuard { public: ... ~ScopeGuard() { if ((on_success && !detector_) || (on_failure && detector_)) func_(); } ... private: UncaughtExceptionDetector detector_; ... };}{}

\end{code}

% EN: The need to use C++17 may present a (hopefully short-term) obstacle to using this technique in programs constrained to older versions of the language. While there is no other standard-compliant, portable way to solve this problem, most modern compilers have ways to get to the uncaught exception counter. This is how it is done in GCC or Clang (the names starting with \texttt{\_\_} are GCC internal types and functions):
\Trans{The need to use C++17 may present a (hopefully short-term) obstacle to using this technique in programs constrained to older versions of the language. While there is no other standard-compliant, portable way to solve this problem, most modern compilers have ways to get to the uncaught exception counter. This is how it is done in GCC or Clang (the names starting with \texttt{\_\_} are GCC internal types and functions):}{}

\begin{code}
% EN: // Example 10b
namespace  cxxabiv1 {
  struct cxa_eh_globals;
  extern "C" cxa_eh_globals* cxa_get_globals() noexcept;
\Trans{// Example 10b namespace cxxabiv1 { struct cxa_eh_globals; extern "C" cxa_eh_globals* cxa_get_globals() noexcept;}{}

}
% EN: class UncaughtExceptionDetector {
  public:
  UncaughtExceptionDetector() :
    count_(uncaught_exceptions()) { }
  operator bool() const noexcept {
    return uncaught_exceptions() > count_;
   }
  private:
  const int count_;
  int uncaught_exceptions() const noexcept {
    return *(reinterpret_cast<int*>(
      static_cast<char*>( static_cast<void*>(
        cxxabiv1::cxa_get_globals())) + sizeof(void*)));
   }
 };
\Trans{class UncaughtExceptionDetector { public: UncaughtExceptionDetector() : count_(uncaught_exceptions()) {} operator bool() const noexcept { return uncaught_exceptions() > count_; } private: const int count_; int uncaught_exceptions() const noexcept { return *(reinterpret_cast<int*>( static_cast<char*>( static_cast<void*>( cxxabiv1::cxa_get_globals())) + sizeof(void*))); } };}{}

\end{code}

% EN: Whether we use the exception-driven ScopeGuard or the explicitly named ScopeGuard (perhaps to handle error codes as well as exceptions), we have accomplished our goals---we now can specify deferred actions that must be taken at the end of a function or any other scope.
\Trans{Whether we use the exception-driven ScopeGuard or the explicitly named ScopeGuard (perhaps to handle error codes as well as exceptions), we have accomplished our goals---we now can specify deferred actions that must be taken at the end of a function or any other scope.}{}

% EN: At the end of this chapter, we will show another implementation of ScopeGuard that can be found in several sources on the web. This implementation deserves some consideration, but you should be aware of the downsides as well.
\Trans{At the end of this chapter, we will show another implementation of ScopeGuard that can be found in several sources on the web. This implementation deserves some consideration, but you should be aware of the downsides as well.}{}

\section{Type-erased ScopeGuard}

% EN: If you search online for a ScopeGuard example, you may chance upon an implementation that uses \texttt{std::function} instead of a class template. The implementation itself is quite simple:
\Trans{If you search online for a ScopeGuard example, you may chance upon an implementation that uses \texttt{std::function} instead of a class template. The implementation itself is quite simple:}{}

\begin{code}
% EN: // Example 11
class ScopeGuard {
  public:
  template <typename Func> ScopeGuard(Func&& func) :
    func_(std::forward<Func>(func)) { }
  ~ScopeGuard() { if (!commit_) func_();  }
  void commit() const noexcept { commit_ = true;  }
  ScopeGuard(ScopeGuard&& other) :
    commit_(other.commit_), func_(std::move(other.func_)) {
    other.commit();
   }
  private:
  mutable bool commit_ = false;
  std::function<void()> func_;
  ScopeGuard& operator=(const ScopeGuard&) = delete;
 };
\Trans{// Example 11 class ScopeGuard { public: template <typename Func> ScopeGuard(Func&& func) : func_(std::forward<Func>(func)) {} ~ScopeGuard() { if (!commit_) func_(); } void commit() const noexcept { commit_ = true; } ScopeGuard(ScopeGuard&& other) : commit_(other.commit_), func_(std::move(other.func_)) { other.commit(); } private: mutable bool commit_ = false; std::function<void()> func_; ScopeGuard& operator=(const ScopeGuard&) = delete; };}{}

\end{code}

% EN: Note that this ScopeGuard is a class, not a class template. It has template constructors that can accept the same lambda expression or another callable object as the other guard. But the variable used to store that expression has the same type no matter what the type of the callable is. That type is \texttt{std::function\textless{}void()\textgreater{}}, a wrapper for any function that takes no arguments and returns nothing. How can a value of any type be stored in an object of some fixed type? That is the magic of type erasure, and we have a whole chapter dedicated to it (\emph{Chapter 6, Understanding Type Erasure}). This non-template ScopeGuard makes the code that uses it simpler (at least in the pre-C++17 version) because there are no types to deduce:
\Trans{Note that this ScopeGuard is a class, not a class template. It has template constructors that can accept the same lambda expression or another callable object as the other guard. But the variable used to store that expression has the same type no matter what the type of the callable is. That type is \texttt{std::function\textless{}void()\textgreater{}}, a wrapper for any function that takes no arguments and returns nothing. How can a value of any type be stored in an object of some fixed type? That is the magic of type erasure, and we have a whole chapter dedicated to it (\emph{Chapter 6, Understanding Type Erasure}). This non-template ScopeGuard makes the code that uses it simpler (at least in the pre-C++17 version) because there are no types to deduce:}{}

\begin{code}
% EN: void Database::insert(const Record& r) {
  S.insert(r);
  ScopeGuard SF([&] { S.finalize();  });
  ScopeGuard SG([&] { S.undo();  });
  I.insert(r);
  SG.commit();
\Trans{void Database::insert(const Record& r) { S.insert(r); ScopeGuard SF([&] { S.finalize(); }); ScopeGuard SG([&] { S.undo(); }); I.insert(r); SG.commit();}{}

}
\end{code}

% EN: There is, however, a serious downside to this approach---a type-erased object has to do a non-trivial amount of computation to achieve its magic. As a minimum, it involves an indirect or a virtual function call, and often some memory is allocated and deallocated as well.
\Trans{There is, however, a serious downside to this approach---a type-erased object has to do a non-trivial amount of computation to achieve its magic. As a minimum, it involves an indirect or a virtual function call, and often some memory is allocated and deallocated as well.}{}

% EN: Using the lessons of \emph{Chapter 6, Understanding Type Erasure}, we can come up with a slightly more efficient type erasure implementation; in particular, we can insist that the on-exit callable fits into the guard's buffer:
\Trans{Using the lessons of \emph{Chapter 6, Understanding Type Erasure}, we can come up with a slightly more efficient type erasure implementation; in particular, we can insist that the on-exit callable fits into the guard's buffer:}{}

\begin{code}
% EN: // Example 11
template <size_t S = 16>
class ScopeGuard : public CommitFlag {
  alignas(8) char space_[S];
  using guard_t = void(*)(void*);
  guard_t guard_ = nullptr;
  template<typename Callable>
  static void invoke(void* callable) {
    (*static_cast<Callable*>(callable))();
   }
  mutable bool commit_ = false;
  public:
  template <typename Callable,
            typename D = std::decay_t<Callable>>
    ScopeGuard(Callable&& callable) :
    guard_(invoke<Callable>) {
    static_assert(sizeof(Callable) <= sizeof(space_));
    ::new(static_cast<void*>(space_))
      D(std::forward<Callable>(callable));
   }
  ScopeGuard(ScopeGuard&& other) = default;
  ~ScopeGuard() { if (!commit_) guard_(space_);  }
 };
\Trans{// Example 11 template <size_t S = 16> class ScopeGuard : public CommitFlag { alignas(8) char space_[S]; using guard_t = void(*)(void*); guard_t guard_ = nullptr; template<typename Callable> static void invoke(void* callable) { (*static_cast<Callable*>(callable))(); } mutable bool commit_ = false; public: template <typename Callable, typename D = std::decay_t<Callable>> ScopeGuard(Callable&& callable) : guard_(invoke<Callable>) { static_assert(sizeof(Callable) <= sizeof(space_)); ::new(static_cast<void*>(space_)) D(std::forward<Callable>(callable)); } ScopeGuard(ScopeGuard&& other) = default; ~ScopeGuard() { if (!commit_) guard_(space_); } };}{}

\end{code}

% EN: We can compare the runtime cost of the type-erased ScopeGuard versus the template \texttt{ScopeGuard} using the Google Benchmark library. The results will depend on what operation we are guarding, of course: for long computations and expensive on-exit actions, a slight difference in \texttt{ScopeGuard}'s run time makes no difference. The difference is going to be more pronounced if the computations in the scope and upon exiting the scope are fast:
\Trans{We can compare the runtime cost of the type-erased ScopeGuard versus the template \texttt{ScopeGuard} using the Google Benchmark library. The results will depend on what operation we are guarding, of course: for long computations and expensive on-exit actions, a slight difference in \texttt{ScopeGuard}'s run time makes no difference. The difference is going to be more pronounced if the computations in the scope and upon exiting the scope are fast:}{}

\begin{code}
% EN: void BM_nodelete(benchmark::State& state) {
  for (auto _ : state) {
    int* p = nullptr;
    ScopeGuardTypeErased::ScopeGuard SG([&] { delete p;  });
    p = rand() < 0 ? new int(42) : nullptr;
   }
  state.SetItemsProcessed(state.iterations());
\Trans{void BM_nodelete(benchmark::State& state) { for (auto _ : state) { int* p = nullptr; ScopeGuardTypeErased::ScopeGuard SG([&] { delete p; }); p = rand() < 0 ? new int(42) : nullptr; } state.SetItemsProcessed(state.iterations());}{}

}
\end{code}

% EN: Note that the memory is never allocated (\texttt{rand()} returns non-negative random numbers) and the pointer \texttt{p} is always \texttt{null}, so we are benchmarking the call to \texttt{rand()} plus the overhead of \texttt{ScopeGuard}. For comparison, we can call \texttt{delete} explicitly, without the guard. The result shows that the template version of the guard has no measurable overhead, while both type-erased implementations have some:
\Trans{Note that the memory is never allocated (\texttt{rand()} returns non-negative random numbers) and the pointer \texttt{p} is always \texttt{null}, so we are benchmarking the call to \texttt{rand()} plus the overhead of \texttt{ScopeGuard}. For comparison, we can call \texttt{delete} explicitly, without the guard. The result shows that the template version of the guard has no measurable overhead, while both type-erased implementations have some:}{}

\begin{code}
% EN: Benchmark                              Time
-------------------------------------------
BM_nodelete_explicit                4.48 ns
BM_nodelete_type_erased             6.29 ns
BM_nodelete_type_erased_fast        5.48 ns
BM_nodelete_template                4.50 ns
\Trans{Benchmark Time ------------------------------------------- BM_nodelete_explicit 4.48 ns BM_nodelete_type_erased 6.29 ns BM_nodelete_type_erased_fast 5.48 ns BM_nodelete_template 4.50 ns}{}

\end{code}

% EN: Our own type-erased version adds about 1 nanosecond to each iteration, and the one based on \texttt{std::function} takes almost twice as long. Benchmarks of this sort are strongly affected by compiler optimizations and can produce very different results in response to even slight changes in the code. For example, let us change the code to always construct the new object:
\Trans{Our own type-erased version adds about 1 nanosecond to each iteration, and the one based on \texttt{std::function} takes almost twice as long. Benchmarks of this sort are strongly affected by compiler optimizations and can produce very different results in response to even slight changes in the code. For example, let us change the code to always construct the new object:}{}

\begin{code}
% EN: void BM_nodelete(benchmark::State& state) {
  for (auto _ : state) {
    int* p = nullptr;
    ScopeGuardTypeErased::ScopeGuard SG([&] { delete p;  });
    p = rand() >= 0 ? new int(42) : nullptr;
   }
  state.SetItemsProcessed(state.iterations());
\Trans{void BM_nodelete(benchmark::State& state) { for (auto _ : state) { int* p = nullptr; ScopeGuardTypeErased::ScopeGuard SG([&] { delete p; }); p = rand() >= 0 ? new int(42) : nullptr; } state.SetItemsProcessed(state.iterations());}{}

}
\end{code}

% EN: Now we call the operator new on each iteration of the loop, so the corresponding deletion must happen as well. This time, the compiler was able to optimize the template \texttt{ScopeGuard} much better than the type-erased one:
\Trans{Now we call the operator new on each iteration of the loop, so the corresponding deletion must happen as well. This time, the compiler was able to optimize the template \texttt{ScopeGuard} much better than the type-erased one:}{}

\begin{code}
% EN: Benchmark                              Time
-------------------------------------------
BM_delete_explicit                  4.54 ns
BM_delete_type_erased               13.4 ns
BM_delete_type_erased_fast          12.7 ns
BM_delete_template                  4.56 ns
\Trans{Benchmark Time ------------------------------------------- BM_delete_explicit 4.54 ns BM_delete_type_erased 13.4 ns BM_delete_type_erased_fast 12.7 ns BM_delete_template 4.56 ns}{}

\end{code}

% EN: Overall, there isn't much of a reason to use type erasure here. The runtime cost may be negligible or significant, but there is usually nothing to gain in exchange. The only advantage of the type-erased version is that the guard itself is always of the same type. But the type of the guard is almost always irrelevant: we create the variable as \texttt{auto} or using constructor template argument deduction, and the only explicit operation we may need to do on the guard is to disarm it. Thus, we never need to write any code that depends on the guard's type. Overall, the template-based ScopeGuard, with or without macros, remains the pattern of choice for automatically releasing resources and performing other actions at the end of a scope.
\Trans{Overall, there isn't much of a reason to use type erasure here. The runtime cost may be negligible or significant, but there is usually nothing to gain in exchange. The only advantage of the type-erased version is that the guard itself is always of the same type. But the type of the guard is almost always irrelevant: we create the variable as \texttt{auto} or using constructor template argument deduction, and the only explicit operation we may need to do on the guard is to disarm it. Thus, we never need to write any code that depends on the guard's type. Overall, the template-based ScopeGuard, with or without macros, remains the pattern of choice for automatically releasing resources and performing other actions at the end of a scope.}{}

\section{Summary}

% EN: In this chapter, we have studied, in detail, one of the best C++ patterns for writing exception-safe and error-safe code. The ScopeGuard pattern allows us to schedule an arbitrary action, a fragment of C++ code, to be executed upon completion of a scope. The scope may be a function, the body of a loop, or just a scope inserted into the program to manage the lifetime of local variables. The actions that are executed to the end may be conditional on the successful completion of the scope, however, that is defined. The ScopeGuard pattern works equally well when success or failure are indicated by return codes or exceptions, although in the latter case we can automatically detect the failure (with return codes, the programmer has to explicitly specify which return values mean success and which do not). We have observed the evolution of the ScopeGuard pattern as more recent language features are used. In its optimal form, ScopeGuard provides a simple declarative way to specify post-conditions and deferred actions, such as cleanup or rollback, in a manner that is trivially composable for any number of actions that need to be committed or undone.
\Trans{In this chapter, we have studied, in detail, one of the best C++ patterns for writing exception-safe and error-safe code. The ScopeGuard pattern allows us to schedule an arbitrary action, a fragment of C++ code, to be executed upon completion of a scope. The scope may be a function, the body of a loop, or just a scope inserted into the program to manage the lifetime of local variables. The actions that are executed to the end may be conditional on the successful completion of the scope, however, that is defined. The ScopeGuard pattern works equally well when success or failure are indicated by return codes or exceptions, although in the latter case we can automatically detect the failure (with return codes, the programmer has to explicitly specify which return values mean success and which do not). We have observed the evolution of the ScopeGuard pattern as more recent language features are used. In its optimal form, ScopeGuard provides a simple declarative way to specify post-conditions and deferred actions, such as cleanup or rollback, in a manner that is trivially composable for any number of actions that need to be committed or undone.}{}

% EN: The next chapter describes another very C++-specific pattern, the Friends Factory, which is a kind of Factory, only instead of objects during the execution of a program, it manufactures functions during its compilation.
\Trans{The next chapter describes another very C++-specific pattern, the Friends Factory, which is a kind of Factory, only instead of objects during the execution of a program, it manufactures functions during its compilation.}{}

\section{Questions}

\begin{enumerate}
\item
% EN:   What is an error-safe or exception-safe, program?
\Trans{What is an error-safe or exception-safe, program?}{}
\item
% EN:   How can we make a routine that performs several related actions be error-safe?
\Trans{How can we make a routine that performs several related actions be error-safe?}{}
\item
% EN:   How does RAII assist in writing error-safe programs?
\Trans{How does RAII assist in writing error-safe programs?}{}
\item
% EN:   How does the ScopeGuard pattern generalize the RAII idiom?
\Trans{How does the ScopeGuard pattern generalize the RAII idiom?}{}
\item
% EN:   How can the program automatically detect when a function exits successfully and when it fails?
\Trans{How can the program automatically detect when a function exits successfully and when it fails?}{}
\item
% EN:   What are the advantages and drawbacks of a type-erased ScopeGuard
\Trans{What are the advantages and drawbacks of a type-erased ScopeGuard}{}
\end{enumerate}

